From 4498a395b3f16cb54fe352322a471c029c2f2b9c Mon Sep 17 00:00:00 2001
From: Chris Packham <judge.packham@gmail.com>
Date: Mon, 8 Aug 2022 20:36:31 +1200
Subject: [PATCH] check_ldrunpath_length

---
 ld/ldelf.c |    4 ++++
 1 file changed, 4 insertions(+)

--- a/ld/ldelf.c
+++ b/ld/ldelf.c
@@ -1127,6 +1127,8 @@ ldelf_handle_dt_needed (struct elf_link_
 		  && command_line.rpath == NULL)
 		{
 		  path = (const char *) getenv ("LD_RUN_PATH");
+		  if (path && *path == '\0')
+		    path = NULL;
 		  if (path
 		      && ldelf_search_needed (path, &n, force,
 					      is_linux, elfsize))
@@ -1802,6 +1804,8 @@ ldelf_before_allocation (char *audit, ch
   rpath = command_line.rpath;
   if (rpath == NULL)
     rpath = (const char *) getenv ("LD_RUN_PATH");
+  if (rpath && *rpath == '\0')
+    rpath = NULL;
 
   for (abfd = link_info.input_bfds; abfd; abfd = abfd->link.next)
     if (bfd_get_flavour (abfd) == bfd_target_elf_flavour)
