diff -durN gcc-4.4.0.orig/libgomp/configure gcc-4.4.0/libgomp/configure
--- gcc-4.4.0.orig/libgomp/configure	2009-04-21 11:08:08.000000000 +0200
+++ gcc-4.4.0/libgomp/configure	2009-05-27 21:38:19.000000000 +0200
@@ -3334,7 +3334,7 @@
 
 # Add -Wall -Werror if we are using GCC.
 if test "x$GCC" = "xyes"; then
-  XCFLAGS="$XCFLAGS -Wall -Werror"
+  XCFLAGS="$XCFLAGS -Wall"
 fi
 
 # Find other programs we need.
