diff -urN uClibc-dist/libc/sysdeps/linux/common/bits/uClibc_touplow.h uClibc/libc/sysdeps/linux/common/bits/uClibc_touplow.h
--- uClibc-dist/libc/sysdeps/linux/common/bits/uClibc_touplow.h   Fri Jun  3 13:30:25 2005
+++ uClibc/libc/sysdeps/linux/common/bits/uClibc_touplow.h   Fri Jun  3 13:30:39 2005
@@ -34,11 +34,11 @@

 /* glibc uses the equivalent of - typedef __int32_t __ctype_touplow_t; */

-typedef __uint16_t __ctype_mask_t;
+typedef __uint32_t __ctype_mask_t;

 #ifdef __UCLIBC_HAS_CTYPE_SIGNED__

-typedef __int16_t __ctype_touplow_t;
+typedef __int32_t __ctype_touplow_t;
 #define __UCLIBC_CTYPE_B_TBL_OFFSET       128
 #define __UCLIBC_CTYPE_TO_TBL_OFFSET      128

