diff -dur strace-4.5.15.orig/net.c strace-4.5.15/net.c
--- strace-4.5.15.orig/net.c	2006-12-13 20:57:23.000000000 +0100
+++ strace-4.5.15/net.c	2007-07-14 19:29:36.000000000 +0200
@@ -800,8 +800,12 @@
 
 #ifdef SOL_PACKET
 static const struct xlat sockpacketoptions[] = {
+#if defined(PACKET_ADD_MEMBERSHIP)
 	{ PACKET_ADD_MEMBERSHIP,	"PACKET_ADD_MEMBERSHIP"	},
+#endif
+#if defined(PACKET_DROP_MEMBERSHIP)
 	{ PACKET_DROP_MEMBERSHIP,	"PACKET_DROP_MEMBERSHIP"},
+#endif
 #if defined(PACKET_RECV_OUTPUT)
 	{ PACKET_RECV_OUTPUT,		"PACKET_RECV_OUTPUT"	},
 #endif
