diff -dur strace-4.5.15.orig/signal.c strace-4.5.15/signal.c
--- strace-4.5.15.orig/signal.c	2007-01-11 23:08:38.000000000 +0100
+++ strace-4.5.15/signal.c	2007-07-14 19:24:43.000000000 +0200
@@ -1440,7 +1440,7 @@
 	  	tcp->u_rval = tcp->u_error = 0;
 		if(tcp->u_arg[0] == 0)
 		  	return 0;
-		tcp->auxstr = sprintsigmask("mask now ", tcp->u_arg[1]);
+		tcp->auxstr = sprintsigmask("mask now ", tcp->u_arg[1], 0);
 		return RVAL_NONE | RVAL_STR;
 	}
 	return 0;
