diff -Naur libelf-0.8.13-orig/configure libelf-0.8.13/configure
--- libelf-0.8.13-orig/configure	2024-05-16 08:46:44.055714647 +1200
+++ libelf-0.8.13/configure	2024-05-16 13:21:21.857127252 +1200
@@ -734,7 +734,7 @@
 #line 735 "configure"
 #include "confdefs.h"
 
-main(){return(0);}
+int main(){return(0);}
 EOF
 if { (eval echo configure:740: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext}; then
   ac_cv_prog_cc_works=yes
@@ -1484,12 +1484,13 @@
 #line 1485 "configure"
 #include "confdefs.h"
 #include <stdio.h>
-main()
+int main()
 {
   FILE *f=fopen("conftestval", "w");
   if (!f) exit(1);
   fprintf(f, "%d\n", sizeof(short));
   exit(0);
+  return 0;
 }
 EOF
 if { (eval echo configure:1496: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext} && (./conftest; exit) 2>/dev/null
@@ -1523,12 +1524,13 @@
 #line 1524 "configure"
 #include "confdefs.h"
 #include <stdio.h>
-main()
+int main()
 {
   FILE *f=fopen("conftestval", "w");
   if (!f) exit(1);
   fprintf(f, "%d\n", sizeof(int));
   exit(0);
+  return 0;
 }
 EOF
 if { (eval echo configure:1535: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext} && (./conftest; exit) 2>/dev/null
@@ -1562,12 +1564,13 @@
 #line 1563 "configure"
 #include "confdefs.h"
 #include <stdio.h>
-main()
+int main()
 {
   FILE *f=fopen("conftestval", "w");
   if (!f) exit(1);
   fprintf(f, "%d\n", sizeof(long));
   exit(0);
+  return 0;
 }
 EOF
 if { (eval echo configure:1574: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext} && (./conftest; exit) 2>/dev/null
@@ -1601,12 +1604,13 @@
 #line 1602 "configure"
 #include "confdefs.h"
 #include <stdio.h>
-main()
+int main()
 {
   FILE *f=fopen("conftestval", "w");
   if (!f) exit(1);
   fprintf(f, "%d\n", sizeof(long long));
   exit(0);
+  return 0;
 }
 EOF
 if { (eval echo configure:1613: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext} && (./conftest; exit) 2>/dev/null
@@ -1641,12 +1645,13 @@
 #line 1642 "configure"
 #include "confdefs.h"
 #include <stdio.h>
-main()
+int main()
 {
   FILE *f=fopen("conftestval", "w");
   if (!f) exit(1);
   fprintf(f, "%d\n", sizeof(__int64));
   exit(0);
+  return 0;
 }
 EOF
 if { (eval echo configure:1653: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext} && (./conftest; exit) 2>/dev/null
@@ -2534,11 +2539,12 @@
 #define memmove(d,s,n) bcopy((s),(d),(n))
 #endif
 extern int strcmp();
-main() {
+int main() {
   char buf[] = "0123456789";
   memmove(buf + 1, buf, 9);
   if (strcmp(buf, "0012345678")) exit(1);
   exit(0);
+  retrun 0;
 }
 EOF
 if { (eval echo configure:2545: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext} && (./conftest; exit) 2>/dev/null
