diff -Naur libelf-0.8.13-orig/configure libelf-0.8.13/configure
--- libelf-0.8.13-orig/configure	2025-02-01 13:58:59.204951646 +1300
+++ libelf-0.8.13/configure	2025-02-01 14:30:18.978354125 +1300
@@ -1114,6 +1114,7 @@
   cat > conftest.$ac_ext <<EOF
 #line 1116 "configure"
 #include "confdefs.h"
+#include <stdlib.h>
 #include <ctype.h>
 #define ISLOWER(c) ('a' <= (c) && (c) <= 'z')
 #define TOUPPER(c) (ISLOWER(c) ? 'A' + ((c) - 'a') : (c))
@@ -1484,6 +1485,7 @@
 #line 1485 "configure"
 #include "confdefs.h"
 #include <stdio.h>
+#include <stdlib.h>
 int main()
 {
   FILE *f=fopen("conftestval", "w");
@@ -1524,6 +1526,7 @@
 #line 1524 "configure"
 #include "confdefs.h"
 #include <stdio.h>
+#include <stdlib.h>
 int main()
 {
   FILE *f=fopen("conftestval", "w");
@@ -1564,6 +1567,7 @@
 #line 1563 "configure"
 #include "confdefs.h"
 #include <stdio.h>
+#include <stdlib.h>
 int main()
 {
   FILE *f=fopen("conftestval", "w");
@@ -1604,6 +1608,7 @@
 #line 1602 "configure"
 #include "confdefs.h"
 #include <stdio.h>
+#include <stdlib.h>
 int main()
 {
   FILE *f=fopen("conftestval", "w");
@@ -1645,6 +1650,7 @@
 #line 1642 "configure"
 #include "confdefs.h"
 #include <stdio.h>
+#include <stdlib.h>
 int main()
 {
   FILE *f=fopen("conftestval", "w");
