diff -dur libelf-0.8.10.orig/configure libelf-0.8.10/configure
--- libelf-0.8.10.orig/configure	2007-09-07 14:08:06.000000000 +0200
+++ libelf-0.8.10/configure	2008-10-23 22:43:19.000000000 +0200
@@ -1595,7 +1595,7 @@
   echo $ac_n "(cached) $ac_c" 1>&6
 else
   if test "$cross_compiling" = yes; then
-  ac_cv_sizeof_long_long=0
+  ac_cv_sizeof_long_long=8
 else
   cat > conftest.$ac_ext <<EOF
 #line 1602 "configure"
diff -dur libelf-0.8.10.orig/configure.in libelf-0.8.10/configure.in
--- libelf-0.8.10.orig/configure.in	2007-09-07 14:07:59.000000000 +0200
+++ libelf-0.8.10/configure.in	2008-10-23 22:43:10.000000000 +0200
@@ -90,7 +90,7 @@
 AC_CHECK_SIZEOF(short,2)
 AC_CHECK_SIZEOF(int,4)
 AC_CHECK_SIZEOF(long,4)
-AC_CHECK_SIZEOF(long long,0)
+AC_CHECK_SIZEOF(long long,8)
 # Windows port
 AC_CHECK_SIZEOF(__int64, 0)
 
