Index: binutils-2.19.51.0.5/ld/ld.texinfo
===================================================================
--- binutils-2.19.51.0.5.orig/ld/ld.texinfo
+++ binutils-2.19.51.0.5/ld/ld.texinfo
@@ -2036,8 +2036,9 @@ This linker can create the new dynamic t
 systems may not understand them. If you specify
 @option{--enable-new-dtags}, the dynamic tags will be created as needed.
 If you specify @option{--disable-new-dtags}, no new dynamic tags will be
-created. By default, the new dynamic tags are not created. Note that
-those options are only available for ELF systems.
+created. On Gentoo, by default, the new dynamic tags are created (this
+differs from upstream behaviour). Note that those options are only
+available for ELF systems.
 
 @kindex --hash-size=@var{number}
 @item --hash-size=@var{number}
