diff -u binutils-2.20.orig/configure binutils-2.20/configure
--- binutils-2.20.orig/configure	2010-01-05 11:15:44.000000000 +0100
+++ binutils-2.20/configure	2010-01-05 12:25:41.000000000 +0100
@@ -3086,7 +3086,7 @@
     noconfigdirs="$noconfigdirs target-libffi target-qthreads"
     libgloss_dir=arm
     ;;
-  arm*-*-linux-gnueabi)
+  arm*-*-linux-*gnueabi)
     noconfigdirs="$noconfigdirs target-qthreads"
     case ${with_newlib} in
       no) noconfigdirs="$noconfigdirs target-newlib target-libgloss"
diff -u binutils-2.20.orig/configure.ac binutils-2.20/configure.ac
--- binutils-2.20.orig/configure.ac	2010-01-05 11:15:44.000000000 +0100
+++ binutils-2.20/configure.ac	2010-01-05 12:25:51.000000000 +0100
@@ -573,7 +573,7 @@
     noconfigdirs="$noconfigdirs target-libffi target-qthreads"
     libgloss_dir=arm
     ;;
-  arm*-*-linux-gnueabi)
+  arm*-*-linux-*gnueabi)
     noconfigdirs="$noconfigdirs target-qthreads"
     case ${with_newlib} in
       no) noconfigdirs="$noconfigdirs target-newlib target-libgloss"
