diff -p -c -u -r1.196 elf32-arm.c
--- a/bfd/elf32-arm.c	22 May 2009 11:58:44 -0000	1.196
+++ b/bfd/elf32-arm.c	12 Jun 2009 13:18:20 -0000
@@ -3175,11 +3175,15 @@ arm_type_of_stub (struct bfd_link_info *
 
 	  /* We have an extra 2-bytes reach because of
 	     the mode change (bit 24 (H) of BLX encoding).  */
+	  /* A stub is needed only if this call is not throught a PLT
+	     entry, because PLT stubs handle mode switching
+	     already.  */
 	  if (branch_offset > (ARM_MAX_FWD_BRANCH_OFFSET + 2)
 	      || (branch_offset < ARM_MAX_BWD_BRANCH_OFFSET)
-	      || ((r_type == R_ARM_CALL) && !globals->use_blx)
-	      || (r_type == R_ARM_JUMP24)
-	      || (r_type == R_ARM_PLT32))
+	      || ( (((r_type == R_ARM_CALL) && !globals->use_blx)
+		    || (r_type == R_ARM_JUMP24)
+		    || (r_type == R_ARM_PLT32))
+		   && !use_plt))
 	    {
 	      stub_type = (info->shared | globals->pic_veneer)
 		/* PIC stubs.  */
