--- linux-4.2/arch/x86/tools/relocs.c.orig	2015-10-20 22:36:03.075546600 +0100
+++ linux-4.2/arch/x86/tools/relocs.c	2015-10-20 22:36:04.934950000 +0100
@@ -187,6 +187,13 @@
 	return name;
 }
 
+/* Hack for Cygwin */
+#if ELF_BITS == 64
+#if !defined(R_X86_64_JUMP_SLOT) && defined(R_X86_64_JMP_SLOT)
+#define R_X86_64_JUMP_SLOT R_X86_64_JMP_SLOT
+#endif
+#endif
+
 static const char *rel_type(unsigned type)
 {
 	static const char *type_name[] = {
