diff -durN gcc-4.4.0.orig/gcc/configure gcc-4.4.0/gcc/configure
--- gcc-4.4.0.orig/gcc/configure	2009-05-27 21:38:08.000000000 +0200
+++ gcc-4.4.0/gcc/configure	2009-05-27 21:38:42.000000000 +0200
@@ -22205,7 +22205,7 @@
 	tls_first_minor=14
 	tls_as_opt="-m64 -Aesame --fatal-warnings"
 	;;
-  sh-*-* | sh[34]-*-*)
+  sh-*-* | sh[34]*-*-*)
     conftest_s='
 	.section ".tdata","awT",@progbits
 foo:	.long	25
diff -durN gcc-4.4.0.orig/gcc/configure.ac gcc-4.4.0/gcc/configure.ac
--- gcc-4.4.0.orig/gcc/configure.ac	2009-05-27 21:38:08.000000000 +0200
+++ gcc-4.4.0/gcc/configure.ac	2009-05-27 21:38:42.000000000 +0200
@@ -2673,7 +2673,7 @@
 	tls_first_minor=14
 	tls_as_opt="-m64 -Aesame --fatal-warnings"
 	;;
-  sh-*-* | sh[34]-*-*)
+  sh-*-* | sh[34]*-*-*)
     conftest_s='
 	.section ".tdata","awT",@progbits
 foo:	.long	25
