diff -ru glibc-2.3.6/elf/Makefile glibc-2.3.6.patched/elf/Makefile
--- glibc-2.3.6/elf/Makefile	2010-05-02 18:58:38.000000000 -0400
+++ glibc-2.3.6.patched/elf/Makefile	2010-05-02 18:54:54.000000000 -0400
@@ -292,7 +292,7 @@
 		  $(LDFLAGS-rtld) -Wl,-z,defs -Wl,--verbose 2>&1 |	\
 		  LC_ALL=C \
 		  sed -e '/^=========/,/^=========/!d;/^=========/d'	\
-		      -e 's/\. = 0 + SIZEOF_HEADERS;/& _begin = . - SIZEOF_HEADERS;/' \
+		      -e 's/\. = .* + SIZEOF_HEADERS;/& _begin = . - SIZEOF_HEADERS;/' \
 		  > $@.lds
 	$(LINK.o) -nostdlib -nostartfiles -shared -o $@			\
 		  $(LDFLAGS-rtld) -Wl,-z,defs $(z-now-$(bind-now))	\
