--- glibc-2.2.2/nss/Makefile.orig	Tue May 20 16:02:47 2003
+++ glibc-2.2.2/nss/Makefile	Tue May 20 16:03:52 2003
@@ -41,6 +41,10 @@
 
 tests			= test-netdb
 
+# Disable optimization when building nss to avoid crash described in
+# http://gcc.gnu.org/ml/gcc-bugs/1998-06/msg00168.html
+default_cflags := -g -O0
+
 include ../Makeconfig
 
 ifeq (yes,$(build-static-nss))
--- glibc-2.2.2/locale/Makefile.orig	Tue May 20 17:29:11 2003
+++ glibc-2.2.2/locale/Makefile	Tue May 20 17:29:39 2003
@@ -66,6 +66,10 @@
 GPERF = gperf
 GPERFFLAGS = -acCgopt -k1,2,5,9,$$ -L ANSI-C
 
+# Disable optimization when building locale or nss to avoid crash described in
+# http://gcc.gnu.org/ml/gcc-bugs/1998-06/msg00168.html
+default_cflags := -g -O0
+
 include ../Rules
 
 programs/%-kw.h: programs/%-kw.gperf
