diff -urN glibc-2.2.5-orig/sysdeps/unix/sysv/linux/i386/sysdep.h glibc-2.2.5/sysdeps/unix/sysv/linux/i386/sysdep.h
--- glibc-2.2.5-orig/sysdeps/unix/sysv/linux/i386/sysdep.h	2001-07-06 06:56:16.000000000 +0200
+++ glibc-2.2.5/sysdeps/unix/sysv/linux/i386/sysdep.h	2004-09-03 12:35:44.000000000 +0200
@@ -36,6 +36,16 @@
 #undef L
 #define L(name)	.L##name
 
+/* This is a kludge to make syscalls.list find these under the names
+ *    pread and pwrite, since some kernel headers define those names
+ *       and some define the *64 names for the same system calls.  */
+#if !defined __NR_pread && defined __NR_pread64
+# define __NR_pread __NR_pread64
+#endif
+#if !defined __NR_pwrite && defined __NR_pwrite64
+# define __NR_pwrite __NR_pwrite64
+#endif
+
 #ifdef __ASSEMBLER__
 
 /* Linux uses a negative return value to indicate syscall errors,
