--- glibc-2.7/sysdeps/unix/sysv/linux/kernel-features.h	2007-08-14 06:21:09.000000000 +0300
+++ glibc-2.7/sysdeps/unix/sysv/linux/kernel-features.h.new	2008-07-23 07:05:20.000000000 +0300
@@ -454,8 +454,7 @@
    and on Alpha just after 2.6.22-rc1.  */
 #if __LINUX_KERNEL_VERSION >= 0x020611 \
     && ((!defined __sh__ && !defined __alpha__) \
-	|| (__LINUX_KERNEL_VERSION >= 0x020613 && defined __sh__) \
-	|| (__LINUX_KERNEL_VERSION >= 0x020617 && defined __alpha__))
+	|| (__LINUX_KERNEL_VERSION >= 0x020613 && defined __sh__))
 # define __ASSUME_ATFCTS	1
 #endif
 
