http://sourceware.org/ml/libc-alpha/2002-10/msg00392.html

diff -durN glibc-2.13.orig/glibc-ports-2.13/sysdeps/mips/fpu_control.h glibc-2.13/glibc-ports-2.13/sysdeps/mips/fpu_control.h
--- glibc-2.13.orig/glibc-ports-2.13/sysdeps/mips/fpu_control.h	2009-05-16 10:36:20.000000000 +0200
+++ glibc-2.13/glibc-ports-2.13/sysdeps/mips/fpu_control.h	2009-11-13 00:51:31.000000000 +0100
@@ -86,7 +86,7 @@
 #define _FPU_RC_UP      0x2
 #define _FPU_RC_DOWN    0x3
 
-#define _FPU_RESERVED 0xfe3c0000  /* Reserved bits in cw */
+#define _FPU_RESERVED 0xfebc0000  /* Reserved bits in cw */
 
 
 /* The fdlibm code requires strict IEEE double precision arithmetic,
diff -durN glibc-2.13.orig/ports/sysdeps/mips/fpu_control.h glibc-2.13/ports/sysdeps/mips/fpu_control.h
