diff -durN duma_2_5_14.orig/Makefile duma_2_5_14/Makefile
--- duma_2_5_14.orig/Makefile	2008-03-13 00:54:10.000000000 +0100
+++ duma_2_5_14/Makefile	2008-06-17 10:28:45.000000000 +0200
@@ -93,10 +93,6 @@
 # also define 'WIN32'
 
 # some defaults:
-CC=gcc
-CXX=g++
-AR=ar
-RANLIB=ranlib
 INSTALL=install
 RM=rm
 RMFORCE=rm -f
@@ -317,7 +313,7 @@
 
 createconf$(EXEPOSTFIX): createconf.o
 	- $(RMFORCE) createconf$(EXEPOSTFIX)
-	$(CC) $(CFLAGS) $(DUMA_OPTIONS) createconf.o -o createconf$(EXEPOSTFIX)
+	$(HOSTCC) $(HOSTCFLAGS) $(DUMA_OPTIONS) createconf.o -o createconf$(EXEPOSTFIX)
 
 tstheap$(EXEPOSTFIX): libduma.a tstheap.o
 	- $(RMFORCE) tstheap$(EXEPOSTFIX)
@@ -378,7 +374,7 @@
 # define rules how to build objects for createconf
 #
 createconf.o:
-	$(CC) $(CFLAGS) $(DUMA_OPTIONS) -c createconf.c -o $@
+	$(HOSTCC) $(HOSTCFLAGS) $(DUMA_OPTIONS) -c createconf.c -o $@
 
 
 #
