diff -urpN mingw-w64-v5.0.1.orig/mingw-w64-tools/genidl/src/genidl_cfg.h mingw-w64-v5.0.1/mingw-w64-tools/genidl/src/genidl_cfg.h
--- mingw-w64-v5.0.1.orig/mingw-w64-tools/genidl/src/genidl_cfg.h	2017-03-10 16:20:09.000000000 -0800
+++ mingw-w64-v5.0.1/mingw-w64-tools/genidl/src/genidl_cfg.h	2017-03-10 16:25:26.000000000 -0800
@@ -26,7 +26,9 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#if defined(HAVE_MALLOC_H)
 #include <malloc.h>
+#endif
 #include <memory.h>
 #include <stdarg.h>
 #include <stdint.h>
diff -urpN mingw-w64-v5.0.1.orig/mingw-w64-tools/genidl/src/genidl_typinfo.c mingw-w64-v5.0.1/mingw-w64-tools/genidl/src/genidl_typinfo.c
--- mingw-w64-v5.0.1.orig/mingw-w64-tools/genidl/src/genidl_typinfo.c	2017-03-10 16:20:09.000000000 -0800
+++ mingw-w64-v5.0.1/mingw-w64-tools/genidl/src/genidl_typinfo.c	2017-03-10 16:25:49.000000000 -0800
@@ -21,7 +21,10 @@
 #include "genidl_cfg.h"
 #include "genidl_typeinfo.h"
 #include "genidl_typinfo.h"
+#if defined(HAVE_MALLOC_H)
 #include <malloc.h>
+#endif
+#include <stdlib.h>
 #include <string.h>
 #include <memory.h>
 
diff -urpN mingw-w64-v5.0.1.orig/mingw-w64-tools/widl/include/pathtools.h mingw-w64-v5.0.1/mingw-w64-tools/widl/include/pathtools.h
--- mingw-w64-v5.0.1.orig/mingw-w64-tools/widl/include/pathtools.h	2017-03-10 16:20:09.000000000 -0800
+++ mingw-w64-v5.0.1/mingw-w64-tools/widl/include/pathtools.h	2017-03-10 16:26:21.000000000 -0800
@@ -12,9 +12,8 @@
 #define PATHTOOLS_H
 
 #include <unistd.h>
-#if defined(__APPLE__)
 #include <stdlib.h>
-#else
+#if defined(HAVE_MALLOC_H)
 #include <malloc.h>
 #endif
 #include <stdio.h>
diff -urpN mingw-w64-v5.0.1.orig/mingw-w64-tools/widl/src/pathtools.c mingw-w64-v5.0.1/mingw-w64-tools/widl/src/pathtools.c
--- mingw-w64-v5.0.1.orig/mingw-w64-tools/widl/src/pathtools.c	2017-03-10 16:20:09.000000000 -0800
+++ mingw-w64-v5.0.1/mingw-w64-tools/widl/src/pathtools.c	2017-03-10 16:26:40.000000000 -0800
@@ -8,9 +8,8 @@
   .email: <mingw.android@gmail.com>.
  */
 
-#if defined(__APPLE__)
 #include <stdlib.h>
-#else
+#if defined(HAVE_MALLOC_H)
 #include <malloc.h>
 #endif
 #include <limits.h>
