diff -urpN strace-4.15.orig/configure strace-4.15/configure
--- strace-4.15.orig/configure	2016-12-14 01:17:44.000000000 -0800
+++ strace-4.15/configure	2017-01-19 22:32:20.167896014 -0800
@@ -8775,7 +8775,7 @@ $as_echo_n "checking for _MIPS_SIM... "
 if ${st_cv__MIPS_SIM+:} false; then :
   $as_echo_n "(cached) " >&6
 else
-  if ac_fn_c_compute_int "$LINENO" "_MIPS_SIM" "st_cv__MIPS_SIM"        "#include <sgidefs.h>"; then :
+  if ac_fn_c_compute_int "$LINENO" "_MIPS_SIM" "st_cv__MIPS_SIM"        "#include <asm/sgidefs.h>"; then :
 
 else
   as_fn_error $? "_MIPS_SIM cannot be determined" "$LINENO" 5
@@ -8792,7 +8792,7 @@ if ${st_cv_mips_abi+:} false; then :
 else
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
-#include <sgidefs.h>
+#include <asm/sgidefs.h>
 int
 main ()
 {
@@ -8806,7 +8806,7 @@ if ac_fn_c_try_compile "$LINENO"; then :
 else
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
-#include <sgidefs.h>
+#include <asm/sgidefs.h>
 int
 main ()
 {
@@ -8820,7 +8820,7 @@ if ac_fn_c_try_compile "$LINENO"; then :
 else
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
-#include <sgidefs.h>
+#include <asm/sgidefs.h>
 int
 main ()
 {
diff -urpN strace-4.15.orig/configure.ac strace-4.15/configure.ac
--- strace-4.15.orig/configure.ac	2016-12-07 02:59:52.000000000 -0800
+++ strace-4.15/configure.ac	2017-01-19 22:32:35.768433437 -0800
@@ -212,20 +212,20 @@ MIPS_ABI=
 if test "$arch" = mips; then
 	AC_CACHE_CHECK([for _MIPS_SIM], [st_cv__MIPS_SIM],
 		       [AC_COMPUTE_INT([st_cv__MIPS_SIM], [_MIPS_SIM],
-				       [#include <sgidefs.h>],
+				       [#include <asm/sgidefs.h>],
 				       [AC_MSG_ERROR([_MIPS_SIM cannot be determined])])])
 
 	AC_CACHE_CHECK([for MIPS ABI], [st_cv_mips_abi],
 		[AC_COMPILE_IFELSE(
-			[AC_LANG_PROGRAM([[#include <sgidefs.h>]],
+			[AC_LANG_PROGRAM([[#include <asm/sgidefs.h>]],
 					 [[int i[_MIPS_SIM == _MIPS_SIM_ABI32 ? 1 : - 1];]])],
 			[st_cv_mips_abi=o32],
 		[AC_COMPILE_IFELSE(
-			[AC_LANG_PROGRAM([[#include <sgidefs.h>]],
+			[AC_LANG_PROGRAM([[#include <asm/sgidefs.h>]],
 					 [[int i[_MIPS_SIM == _MIPS_SIM_NABI32 ? 1 : - 1];]])],
 			[st_cv_mips_abi=n32],
 		[AC_COMPILE_IFELSE(
-			[AC_LANG_PROGRAM([[#include <sgidefs.h>]],
+			[AC_LANG_PROGRAM([[#include <asm/sgidefs.h>]],
 					 [[int i[_MIPS_SIM == _MIPS_SIM_ABI64 ? 1 : - 1];]])],
 			[st_cv_mips_abi=n64],
 			[st_cv_mips_abi=unknown])])])])
