diff -durN glibc-2.7.orig/Makerules glibc-2.7/Makerules
--- glibc-2.7.orig/Makerules	2007-08-26 04:18:03.000000000 +0200
+++ glibc-2.7/Makerules	2008-07-28 15:17:27.000000000 +0200
@@ -1203,6 +1203,9 @@
 # The include magic above causes those files to use this variable for flags.
 CPPFLAGS-nonlib = -DNOT_IN_libc=1
 
+ifneq ($(OBJDUMP_FOR_HOST),)
+OBJDUMP = $(OBJDUMP_FOR_HOST)
+endif
 
 ifeq ($(versioning),yes)
 # Generate normalized lists of symbols, versions, and data sizes.
