--- glibc-2.7/sysdeps/unix/sysv/linux/sh/sysdep.S.orig	2005-12-30 22:16:43.000000000 +0000
+++ glibc-2.7/sysdeps/unix/sysv/linux/sh/sysdep.S	2008-08-19 22:33:14.000000000 +0100
@@ -32,3 +32,13 @@
 
 #define __syscall_error __syscall_error_1
 #include <sysdeps/unix/sh/sysdep.S>
+
+	.data
+	.align 3
+	.globl ___fpscr_values
+	.type ___fpscr_values, @object
+	.size ___fpscr_values, 8
+___fpscr_values:
+	.long	0
+	.long	0x80000
+weak_alias (___fpscr_values, __fpscr_values)
--- glibc-2.7/sysdeps/unix/sysv/linux/sh/Versions.orig	2003-09-01 05:05:09.000000000 +0100
+++ glibc-2.7/sysdeps/unix/sysv/linux/sh/Versions	2008-08-19 22:33:14.000000000 +0100
@@ -2,6 +2,7 @@
   GLIBC_2.2 {
     # functions used in other libraries
     __xstat64; __fxstat64; __lxstat64;
+    __fpscr_values;
 
     # a*
     alphasort64;
