diff -durN glibc-2.7.orig/sysdeps/unix/sysv/linux/sh/sys/procfs.h glibc-2.7/sysdeps/unix/sysv/linux/sh/sys/procfs.h
--- glibc-2.7.orig/sysdeps/unix/sysv/linux/sh/sys/procfs.h	2008-10-22 21:23:32.000000000 +0200
+++ glibc-2.7/sysdeps/unix/sysv/linux/sh/sys/procfs.h	2008-10-22 21:41:54.000000000 +0200
@@ -29,6 +29,7 @@
 #include <sys/types.h>
 #include <sys/ucontext.h>
 #include <sys/user.h>
+#include <asm/ptrace.h>
 
 __BEGIN_DECLS
 
diff -durN glibc-2.7.orig/sysdeps/unix/sysv/linux/sh/sys/user.h glibc-2.7/sysdeps/unix/sysv/linux/sh/sys/user.h
--- glibc-2.7.orig/sysdeps/unix/sysv/linux/sh/sys/user.h	2008-10-22 21:23:32.000000000 +0200
+++ glibc-2.7/sysdeps/unix/sysv/linux/sh/sys/user.h	2008-10-22 21:33:34.000000000 +0200
@@ -21,8 +21,6 @@
 
 #include <features.h>
 
-#include <asm/user.h>
-
 #undef start_thread
 
 #endif  /* sys/user.h */
