diff -durN ltrace-0.4.orig/Makefile.in ltrace-0.4/Makefile.in
--- ltrace-0.4.orig/Makefile.in	2008-10-26 15:34:21.000000000 +0100
+++ ltrace-0.4/Makefile.in	2008-10-26 15:37:40.000000000 +0100
@@ -5,6 +5,7 @@
 OS		:= @HOST_OS@
 ARCH		:= $(shell echo "@HOST_ARCH@" |sed -e s/i.86/i386/      \
 						   -e s/sun4u/sparc64/  \
+						   -e s/alpha.*/alpha/  \
 						   -e s/arm.*/arm/      \
 						   -e s/sa110/arm/      \
 						   -e s/ppc64/ppc/      \
