From 259dc70b4b3bcfa2457650d7e07eba608b748cb1 Mon Sep 17 00:00:00 2001
From: toriningen <toriningen.me@gmail.com>
Date: Sun, 22 Dec 2024 16:07:09 +0200
Subject: [PATCH] Enable install of .exe

---
 Makefile | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index d8ebc4f..3c4e4a6 100644
--- a/Makefile
+++ b/Makefile
@@ -199,7 +199,13 @@ endif
 install-bin: all $(SCRIPTS)
 	@$(VECHO) INSTALL-BIN
 	$(INSTALL) -d $(DESTDIR)$(BINDIR)
-	$(INSTALL_PROGRAM) $(BIN) $(DESTDIR)$(BINDIR)
+	for bin in $(BIN); do \
+	  if [ -f "$$bin" ]; then \
+	    $(INSTALL_PROGRAM) $$bin $(DESTDIR)$(BINDIR); \
+	  elif [ -f "$$bin.exe" ]; then \
+	    $(INSTALL_PROGRAM) $$bin.exe $(DESTDIR)$(BINDIR); \
+	  fi; \
+	done
 	$(INSTALL_SCRIPT) $(SCRIPTS) $(DESTDIR)$(BINDIR)
 
 install-lib: all
-- 
2.47.1

