--- glibc-2.3.2/stdlib/abs.c.old	2004-07-18 12:08:43.000000000 -0700
+++ glibc-2.3.2/stdlib/abs.c	2004-07-18 12:09:18.000000000 -0700
@@ -16,6 +16,8 @@
    Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
    02111-1307 USA.  */
 
+#define __NO_INLINE__ 1	/* avoid http://gcc.gnu.org/PR14096 */
+
 #include <stdlib.h>
 
 #undef	abs
--- glibc-2.3.2/stdlib/atoi.c.old	2004-07-18 12:08:17.000000000 -0700
+++ glibc-2.3.2/stdlib/atoi.c	2004-07-18 12:08:27.000000000 -0700
@@ -16,6 +16,8 @@
    Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
    02111-1307 USA.  */
 
+#define __NO_INLINE__ 1	/* avoid http://gcc.gnu.org/PR14096 */
+
 #include <stdlib.h>
 
 #undef	atoi
--- glibc-2.3.2/stdlib/atof.c.old	2004-07-18 12:06:09.000000000 -0700
+++ glibc-2.3.2/stdlib/atof.c	2004-07-18 12:05:31.000000000 -0700
@@ -16,6 +16,8 @@
    Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
    02111-1307 USA.  */
 
+#define __NO_INLINE__ 1	/* avoid http://gcc.gnu.org/PR14096 */
+
 #include <stdlib.h>
 
 #undef	atof
--- glibc-2.3.2/stdlib/atol.c.old	2004-07-18 12:08:49.000000000 -0700
+++ glibc-2.3.2/stdlib/atol.c	2004-07-18 12:09:40.000000000 -0700
@@ -16,6 +16,8 @@
    Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
    02111-1307 USA.  */
 
+#define __NO_INLINE__ 1	/* avoid http://gcc.gnu.org/PR14096 */
+
 #include <stdlib.h>
 
 #undef	atol
--- glibc-2.3.2/stdlib/atoll.c.old	2004-07-18 12:08:55.000000000 -0700
+++ glibc-2.3.2/stdlib/atoll.c	2004-07-18 12:09:59.000000000 -0700
@@ -16,6 +16,8 @@
    Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
    02111-1307 USA.  */
 
+#define __NO_INLINE__ 1	/* avoid http://gcc.gnu.org/PR14096 */
+
 #include <stdlib.h>
 
 #undef	atoll
