--- gcc-4.2.1-stock/gcc/config/alpha/linux-unwind.h	2005-06-24 20:22:41.000000000 -0500
+++ gcc-4.2.1/gcc/config/alpha/linux-unwind.h	2007-09-01 20:21:34.000000000 -0500
@@ -29,6 +29,9 @@ Boston, MA 02110-1301, USA.  */
 /* Do code reading to identify a signal frame, and set the frame
    state data appropriately.  See unwind-dw2.c for the structs.  */
 
+/* Don't use this if inhibit_libc is set
+   The build for this target will fail trying to include missing headers */
+#ifndef inhibit_libc
 #include <signal.h>
 #include <sys/ucontext.h>
 
@@ -80,3 +83,4 @@ alpha_fallback_frame_state (struct _Unwi
   fs->retaddr_column = 64;
   return _URC_NO_REASON;
 }
+#endif /* inhibit_libc */
