diff -urpN newlib-2.5.0.20170323.orig/config/mt-d30v newlib-2.5.0.20170323/config/mt-d30v
--- newlib-2.5.0.20170323.orig/config/mt-d30v	2017-04-11 00:51:28.180768874 -0700
+++ newlib-2.5.0.20170323/config/mt-d30v	2017-04-11 00:55:51.138626731 -0700
@@ -1,4 +1,4 @@
 # Build libraries optimizing for space, not speed.
 # Turn off warnings about symbols named the same as registers
- CFLAGS_FOR_TARGET = -g -Os -Wa,-C
- CXXFLAGS_FOR_TARGET = -g -Os -Wa,-C
+ CFLAGS_FOR_TARGET += -g -Os -Wa,-C
+ CXXFLAGS_FOR_TARGET += -g -Os -Wa,-C
