diff -urN gettext-0.19.6.orig/gettext-tools/gnulib-lib/execute.c gettext-0.19.6/gettext-tools/gnulib-lib/execute.c
--- gettext-0.19.6.orig/gettext-tools/gnulib-lib/execute.c	2015-10-24 16:54:39.310372500 +0100
+++ gettext-0.19.6/gettext-tools/gnulib-lib/execute.c	2015-10-24 16:55:35.261954100 +0100
@@ -52,7 +52,7 @@
    __cygwin_environ variable on cygwin64:
    <https://cygwin.com/ml/cygwin/2013-06/msg00228.html>.  */
 #if defined __CYGWIN__ && defined __x86_64__
-extern DLL_VARIABLE char **environ;
+extern __attribute__((dllimport)) char **environ;
 #endif
 
 
diff -urN gettext-0.19.6.orig/gettext-tools/gnulib-lib/spawn-pipe.c gettext-0.19.6/gettext-tools/gnulib-lib/spawn-pipe.c
--- gettext-0.19.6.orig/gettext-tools/gnulib-lib/spawn-pipe.c	2015-09-11 04:03:56.000000000 +0100
+++ gettext-0.19.6/gettext-tools/gnulib-lib/spawn-pipe.c	2015-10-24 16:56:00.885918100 +0100
@@ -52,7 +52,7 @@
    __cygwin_environ variable on cygwin64:
    <https://cygwin.com/ml/cygwin/2013-06/msg00228.html>.  */
 #if defined __CYGWIN__ && defined __x86_64__
-extern DLL_VARIABLE char **environ;
+extern __attribute__((dllimport)) char **environ;
 #endif
 
 
diff -urN gettext-0.19.6.orig/gettext-tools/gnulib-tests/test-environ.c gettext-0.19.6/gettext-tools/gnulib-tests/test-environ.c
--- gettext-0.19.6.orig/gettext-tools/gnulib-tests/test-environ.c	2015-09-11 04:03:58.000000000 +0100
+++ gettext-0.19.6/gettext-tools/gnulib-tests/test-environ.c	2015-10-24 16:56:51.519198900 +0100
@@ -26,7 +26,7 @@
    __cygwin_environ variable on cygwin64:
    <https://cygwin.com/ml/cygwin/2013-06/msg00228.html>.  */
 #if defined __CYGWIN__ && defined __x86_64__
-extern DLL_VARIABLE char **environ;
+extern __attribute__((dllimport)) char **environ;
 #endif
 
 int
diff -urN gettext-0.19.6.orig/gnulib-local/lib/execute.c.diff gettext-0.19.6/gnulib-local/lib/execute.c.diff
--- gettext-0.19.6.orig/gnulib-local/lib/execute.c.diff	2015-08-21 08:18:28.000000000 +0100
+++ gettext-0.19.6/gnulib-local/lib/execute.c.diff	2015-10-24 16:58:17.182285000 +0100
@@ -10,7 +10,7 @@
 +   __cygwin_environ variable on cygwin64:
 +   <https://cygwin.com/ml/cygwin/2013-06/msg00228.html>.  */
 +#if defined __CYGWIN__ && defined __x86_64__
-+extern DLL_VARIABLE char **environ;
++extern __attribute__((dllimport)) char **environ;
 +#endif
 +
  
diff -urN gettext-0.19.6.orig/gnulib-local/lib/spawn-pipe.c.diff gettext-0.19.6/gnulib-local/lib/spawn-pipe.c.diff
--- gettext-0.19.6.orig/gnulib-local/lib/spawn-pipe.c.diff	2015-08-21 08:18:28.000000000 +0100
+++ gettext-0.19.6/gnulib-local/lib/spawn-pipe.c.diff	2015-10-24 16:58:11.741447300 +0100
@@ -10,7 +10,7 @@
 +   __cygwin_environ variable on cygwin64:
 +   <https://cygwin.com/ml/cygwin/2013-06/msg00228.html>.  */
 +#if defined __CYGWIN__ && defined __x86_64__
-+extern DLL_VARIABLE char **environ;
++extern __attribute__((dllimport)) char **environ;
 +#endif
 +
  
diff -urN gettext-0.19.6.orig/gnulib-local/tests/test-environ.c.diff gettext-0.19.6/gnulib-local/tests/test-environ.c.diff
--- gettext-0.19.6.orig/gnulib-local/tests/test-environ.c.diff	2015-08-21 08:18:28.000000000 +0100
+++ gettext-0.19.6/gnulib-local/tests/test-environ.c.diff	2015-10-24 16:58:35.915765100 +0100
@@ -10,7 +10,7 @@
 +   __cygwin_environ variable on cygwin64:
 +   <https://cygwin.com/ml/cygwin/2013-06/msg00228.html>.  */
 +#if defined __CYGWIN__ && defined __x86_64__
-+extern DLL_VARIABLE char **environ;
++extern __attribute__((dllimport)) char **environ;
 +#endif
 +
  int
