--- glibc-2.3.3/locale/weight.h.old	2003-06-11 14:52:12.000000000 -0700
+++ glibc-2.3.3/locale/weight.h	2005-03-15 12:05:37.000000000 -0800
@@ -18,8 +18,14 @@
    02111-1307 USA.  */
 
 /* Find index of weight.  */
+#if __GNUC__ >= 4
+auto inline int32_t
+#else
 static inline int32_t
+#endif
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 2)
 __attribute ((always_inline))
+#endif
 findidx (const unsigned char **cpp)
 {
   int_fast32_t i = table[*(*cpp)++];
--- glibc-2.3.3/locale/weightwc.h.old	2003-06-13 13:44:58.000000000 -0700
+++ glibc-2.3.3/locale/weightwc.h	2005-03-15 12:06:46.000000000 -0800
@@ -18,8 +18,14 @@
    02111-1307 USA.  */
 
 /* Find index of weight.  */
+#if __GNUC__ >= 4
+auto inline int32_t
+#else
 static inline int32_t
+#endif
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 2)
 __attribute ((always_inline))
+#endif
 findidx (const wint_t **cpp)
 {
   int32_t i;
