diff -dur uClibc-0.9.29.orig/extra/Configs/Config.mips uClibc-0.9.29/extra/Configs/Config.mips
--- uClibc-0.9.29.orig/extra/Configs/Config.mips	2007-03-16 20:38:14.000000000 +0100
+++ uClibc-0.9.29/extra/Configs/Config.mips	2007-05-22 19:30:43.000000000 +0200
@@ -71,4 +71,16 @@
 config CONFIG_MIPS_ISA_MIPS64
 	bool "MIPS64"
 
+config CONFIG_MIPS_ISA_CUSTOM
+	bool "Custom"
+
 endchoice
+
+config CONFIG_MIPS_CPU_CFLAGS_CUSTOM
+	string
+	prompt "Custon ISA"
+	depends on CONFIG_MIPS_ISA_CUSTOM
+	default ""
+	help
+	  Enter your custom ISA here (eg: lx4189!).
+
diff -dur uClibc-0.9.29.orig/Rules.mak uClibc-0.9.29/Rules.mak
--- uClibc-0.9.29.orig/Rules.mak	2007-04-17 15:34:11.000000000 +0200
+++ uClibc-0.9.29/Rules.mak	2007-05-22 19:31:48.000000000 +0200
@@ -234,6 +234,7 @@
 	CPU_CFLAGS-$(CONFIG_MIPS_N64_ABI)+=-mabi=64
 	CPU_CFLAGS-$(CONFIG_MIPS_O32_ABI)+=-mabi=32
 	CPU_CFLAGS-$(CONFIG_MIPS_N32_ABI)+=-mabi=n32
+	CPU_CFLAGS-$(CONFIG_MIPS_ISA_CUSTOM)+=-march=$(CONFIG_MIPS_CPU_CFLAGS_CUSTOM)
 endif
 
 ifeq ($(TARGET_ARCH),nios)
