diff -dur strace-4.5.16.orig/process.c strace-4.5.16/process.c
--- strace-4.5.16.orig/process.c	2007-01-11 23:08:38.000000000 +0100
+++ strace-4.5.16/process.c	2007-07-14 19:19:58.000000000 +0200
@@ -2685,7 +2685,6 @@
        { 4*REG_GBR,            "4*REG_GBR"                             },
        { 4*REG_MACH,           "4*REG_MACH"                            },
        { 4*REG_MACL,           "4*REG_MACL"                            },
-       { 4*REG_SYSCALL,        "4*REG_SYSCALL"                         },
        { 4*REG_FPUL,           "4*REG_FPUL"                            },
        { 4*REG_FPREG0,         "4*REG_FPREG0"                          },
        { 4*(REG_FPREG0+1),     "4*REG_FPREG1"                          },
