diff -ur binutils-2.22.org/bfd/bfd-in2.h binutils-2.22/bfd/bfd-in2.h
--- binutils-2.22.org/bfd/bfd-in2.h	2011-09-16 03:15:18.000000000 +0200
+++ binutils-2.22/bfd/bfd-in2.h	2012-11-03 18:25:20.365668349 +0100
@@ -32,6 +32,8 @@
 #ifndef __BFD_H_SEEN__
 #define __BFD_H_SEEN__
 
+#include <sys/stat.h>
+
 #ifdef __cplusplus
 extern "C" {
 #endif
@@ -308,8 +310,6 @@
    && (sec)->sec_info_type != ELF_INFO_TYPE_MERGE		\
    && (sec)->sec_info_type != ELF_INFO_TYPE_JUST_SYMS)
 
-/* Forward define.  */
-struct stat;
 
 typedef enum bfd_print_symbol
 {
diff -ur binutils-2.22.org/bfd/bfd-in.h binutils-2.22/bfd/bfd-in.h
--- binutils-2.22.org/bfd/bfd-in.h	2011-09-16 03:15:18.000000000 +0200
+++ binutils-2.22/bfd/bfd-in.h	2012-11-03 18:24:43.397783645 +0100
@@ -25,6 +25,8 @@
 #ifndef __BFD_H_SEEN__
 #define __BFD_H_SEEN__
 
+#include <sys/stat.h>
+
 #ifdef __cplusplus
 extern "C" {
 #endif
@@ -301,8 +303,6 @@
    && (sec)->sec_info_type != ELF_INFO_TYPE_MERGE		\
    && (sec)->sec_info_type != ELF_INFO_TYPE_JUST_SYMS)
 
-/* Forward define.  */
-struct stat;
 
 typedef enum bfd_print_symbol
 {
