---
 sysdeps/linux-gnu/mips/plt.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/sysdeps/linux-gnu/mips/plt.c
+++ b/sysdeps/linux-gnu/mips/plt.c
@@ -356,8 +356,8 @@
 
 	name = strdup(a_name);
 	if (name == NULL) {
-		fprintf(stderr, "%s: failed %s(%#llx): %s\n", __func__,
-			name, addr, strerror(errno));
+		fprintf(stderr, "%s: failed (%#llx): %s\n", __func__,
+			addr, strerror(errno));
 		goto fail;
 	}
 
