diff -Naur ltrace-0.7.3-orig/sysdeps/linux-gnu/mips/plt.c ltrace-0.7.3/sysdeps/linux-gnu/mips/plt.c
--- ltrace-0.7.3-orig/sysdeps/linux-gnu/mips/plt.c	2013-09-17 11:04:28.000000000 +1200
+++ ltrace-0.7.3/sysdeps/linux-gnu/mips/plt.c	2021-04-14 22:05:45.950413823 +1200
@@ -356,8 +356,8 @@
 
 	name = strdup(a_name);
 	if (name == NULL) {
-		fprintf(stderr, "%s: failed %s(%#llx): %s\n", __func__,
-			name, addr, strerror(errno));
+		fprintf(stderr, "%s: failed (%#llx): %s\n", __func__,
+			addr, strerror(errno));
 		goto fail;
 	}
 
