diff -durN glibc-2.13.orig/elf/dl-lookup.c glibc-2.13/elf/dl-lookup.c
--- glibc-2.13.orig/elf/dl-lookup.c	2009-03-30 23:14:32.000000000 +0200
+++ glibc-2.13/elf/dl-lookup.c	2009-11-13 00:51:36.000000000 +0100
@@ -301,6 +301,12 @@
 	      /* FALLTHROUGH */
 	    case STB_GLOBAL:
	    success:
+#ifdef __mips__
+	      /* HACK: MIPS marks its lazy evaluation stubs with SHN_UNDEF
+		 symbols, we skip them. */
+	      if (sym->st_shndx == SHN_UNDEF)
+		break;
+#endif
 	      /* Global definition.  Just what we need.  */
 	      result->s = sym;
 	      result->m = (struct link_map *) map;
