From be5d482de0099012288e617af0583772fac21714 Mon Sep 17 00:00:00 2001
From: Keith Packard <keithp@keithp.com>
Date: Tue, 23 Aug 2022 22:12:06 -0700
Subject: [PATCH] Add newlib and picolibc as default C library choices

Signed-off-by: Keith Packard <keithp@keithp.com>
---
 gcc/config.gcc |   11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -694,7 +694,7 @@ case ${target} in
 esac
 
 # Common C libraries.
-tm_defines="$tm_defines LIBC_GLIBC=1 LIBC_UCLIBC=2 LIBC_BIONIC=3 LIBC_MUSL=4"
+tm_defines="$tm_defines LIBC_GLIBC=1 LIBC_UCLIBC=2 LIBC_BIONIC=3 LIBC_MUSL=4 LIBC_NEWLIB=5 LIBC_PICOLIBC=6"
 
 default_libc=""
 
@@ -6178,6 +6178,15 @@ bionic)
 musl)
     default_libc=LIBC_MUSL
     ;;
+newlib)
+    # Newlib configurations don't set the DEFAULT_LIBC variable, so
+    # avoid changing those by allowing --with-default-libc=newlib but
+    # not actually setting the DEFAULT_LIBC variable.
+    default_libc=
+    ;;
+picolibc)
+    default_libc=LIBC_PICOLIBC
+    ;;
 "")
     ;;
 *)
