From 1a6a99e9d9cf5f4672bf2b8a96df28ed4e6f0fb3 Mon Sep 17 00:00:00 2001
From: Ahmed ARIF <arif.ing@outlook.com>
Date: Sun, 21 Sep 2025 16:35:43 +0200
Subject: [PATCH] Apply 0021-aarch64-Bypass-some-warnings-in-MinGW

---
 libgcc/config/aarch64/t-mingw | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libgcc/config/aarch64/t-mingw b/libgcc/config/aarch64/t-mingw
index a72a2f701..cff4f8cea 100644
--- a/libgcc/config/aarch64/t-mingw
+++ b/libgcc/config/aarch64/t-mingw
@@ -1,2 +1,2 @@
 # Bypass hidden attribute warnings in MinGW until it is implemented
-LIBGCC2_CFLAGS += -Wno-error=attributes
+LIBGCC2_CFLAGS += -Wno-error=attributes -Wno-error=missing-prototypes
-- 
2.48.1

