diff -durN gcc-3.4.6.orig/libstdc++-v3/include/ext/rope gcc-3.4.6/libstdc++-v3/include/ext/rope
--- gcc-3.4.6.orig/libstdc++-v3/include/ext/rope	2005-01-16 23:08:43.000000000 +0100
+++ gcc-3.4.6/libstdc++-v3/include/ext/rope	2007-08-15 22:53:37.000000000 +0200
@@ -58,6 +58,9 @@
 #include <bits/allocator.h>
 #include <ext/hash_fun.h>
 
+/* cope w/ index defined as macro, SuSv3 proposal */
+#undef index
+
 # ifdef __GC
 #   define __GC_CONST const
 # else
diff -durN gcc-3.4.6.orig/libstdc++-v3/include/ext/ropeimpl.h gcc-3.4.6/libstdc++-v3/include/ext/ropeimpl.h
--- gcc-3.4.6.orig/libstdc++-v3/include/ext/ropeimpl.h	2004-09-08 15:36:04.000000000 +0200
+++ gcc-3.4.6/libstdc++-v3/include/ext/ropeimpl.h	2007-08-15 22:53:37.000000000 +0200
@@ -53,6 +53,9 @@
 #include <ext/memory> // For uninitialized_copy_n
 #include <ext/numeric> // For power
 
+/* cope w/ index defined as macro, SuSv3 proposal */
+#undef index
+
 namespace __gnu_cxx
 {
   using std::size_t;
