diff -durN gcc-3.4.6.orig/libstdc++-v3/fragment.am gcc-3.4.6/libstdc++-v3/fragment.am
--- gcc-3.4.6.orig/libstdc++-v3/fragment.am	2003-08-27 02:38:28.000000000 +0200
+++ gcc-3.4.6/libstdc++-v3/fragment.am	2007-08-15 22:59:09.000000000 +0200
@@ -18,7 +18,7 @@
 	$(WARN_FLAGS) $(WERROR) -fdiagnostics-show-location=once
 
 # -I/-D flags to pass when compiling.
-AM_CPPFLAGS = $(GLIBCXX_INCLUDES)
+AM_CPPFLAGS = $(GLIBCXX_INCLUDES) -I$(toplevel_srcdir)/include
 
 
 
diff -durN gcc-3.4.6.orig/libstdc++-v3/libmath/Makefile.am gcc-3.4.6/libstdc++-v3/libmath/Makefile.am
--- gcc-3.4.6.orig/libstdc++-v3/libmath/Makefile.am	2003-08-27 23:29:43.000000000 +0200
+++ gcc-3.4.6/libstdc++-v3/libmath/Makefile.am	2007-08-15 22:59:09.000000000 +0200
@@ -32,7 +32,7 @@
 
 libmath_la_SOURCES = stubs.c
 
-AM_CPPFLAGS = $(CANADIAN_INCLUDES)
+AM_CPPFLAGS = $(CANADIAN_INCLUDES) -I$(toplevel_srcdir)/include
 
 # Only compiling "C" sources in this directory.
 LIBTOOL = @LIBTOOL@ --tag CC
