diff -durN gcc-3.4.6.orig/gcc/config/sh/sh.h gcc-3.4.6/gcc/config/sh/sh.h
--- gcc-3.4.6.orig/gcc/config/sh/sh.h	2006-01-12 02:20:32.000000000 +0100
+++ gcc-3.4.6/gcc/config/sh/sh.h	2007-08-15 23:03:09.000000000 +0200
@@ -422,6 +422,10 @@
 do {									\
   if (LEVEL)								\
     flag_omit_frame_pointer = -1;					\
+  if (LEVEL <= 2)							\
+    {									\
+      flag_reorder_blocks = 0;						\
+    }									\
   if (SIZE)								\
     target_flags |= SPACE_BIT;						\
   if (TARGET_SHMEDIA && LEVEL > 1)					\
