diff -durN gcc-3.4.6.orig/gcc/config/rs6000/linux-unwind.h gcc-3.4.6/gcc/config/rs6000/linux-unwind.h
--- gcc-3.4.6.orig/gcc/config/rs6000/linux-unwind.h	2005-03-17 11:41:18.000000000 +0100
+++ gcc-3.4.6/gcc/config/rs6000/linux-unwind.h	2007-08-15 23:07:00.000000000 +0200
@@ -32,6 +32,7 @@
    these structs elsewhere;  Many fields are missing, particularly
    from the end of the structures.  */
 
+#ifndef inhibit_libc
 struct gcc_vregs
 {
   __attribute__ ((vector_size (16))) int vr[32];
@@ -320,3 +321,4 @@
 									\
     goto SUCCESS;							\
   } while (0)
+#endif
