---
 gdb/event-loop.h |    4 ++++
 1 file changed, 4 insertions(+)

--- a/gdb/event-loop.h
+++ b/gdb/event-loop.h
@@ -66,6 +66,8 @@
    infinite loop.
 
    Corollary tasks are the creation and deletion of event sources.  */
+#ifndef _EVENT_LOOP_H_
+#define _EVENT_LOOP_H_
 
 typedef void *gdb_client_data;
 struct async_signal_handler;
@@ -127,3 +129,5 @@
 extern void clear_async_event_handler (struct async_event_handler *handler);
 
 extern void initialize_async_signal_handlers (void);
+
+#endif
