

diff -durN binutils-2.20.1.orig/ld/emulparams/elf_i386.sh binutils-2.20.1/ld/emulparams/elf_i386.sh
--- binutils-2.20.1.orig/ld/emulparams/elf_i386.sh	2009-07-29 16:59:23.000000000 +0200
+++ binutils-2.20.1/ld/emulparams/elf_i386.sh	2010-08-17 19:32:12.000000000 +0200
@@ -13,3 +13,13 @@
 NO_SMALL_DATA=yes
 SEPARATE_GOTPLT=12
 IREL_IN_PLT=
+
+# In Gentoo, we install 32bit libs into /lib32 in an
+# ABI setup with amd64/x86
+case "$target" in
+  x86_64*-linux*)
+    case "$EMULATION_NAME" in
+      *i386*) LIBPATH_SUFFIX=32  ;;
+    esac
+  ;;
+esac
