

diff -durN binutils-2.20.1.orig/ld/ldmain.c binutils-2.20.1/ld/ldmain.c
--- binutils-2.20.1.orig/ld/ldmain.c	2010-08-17 19:32:15.000000000 +0200
+++ binutils-2.20.1/ld/ldmain.c	2010-08-17 19:32:17.000000000 +0200
@@ -266,6 +266,9 @@
   link_info.combreloc = TRUE;
   link_info.strip_discarded = TRUE;
   link_info.emit_hash = TRUE;
+#ifndef __mips__
+  link_info.emit_gnu_hash = TRUE;
+#endif
   link_info.callbacks = &link_callbacks;
   link_info.input_bfds_tail = &link_info.input_bfds;
   /* SVR4 linkers seem to set DT_INIT and DT_FINI based on magic _init
