

diff -durN binutils-2.20.1.orig/libiberty/Makefile.in binutils-2.20.1/libiberty/Makefile.in
--- binutils-2.20.1.orig/libiberty/Makefile.in	2009-08-23 21:03:58.000000000 +0200
+++ binutils-2.20.1/libiberty/Makefile.in	2010-08-17 19:32:19.000000000 +0200
@@ -227,6 +227,7 @@
 	  $(AR) $(AR_FLAGS) $(TARGETLIB) \
 	    $(REQUIRED_OFILES) $(EXTRA_OFILES) $(LIBOBJS); \
 	  $(RANLIB) $(TARGETLIB); \
+	  cp $(TARGETLIB) ../ ; \
 	  cd ..; \
 	else true; fi
 
