diff -u binutils-2.20.orig/configure binutils-2.20/configure

diff -durN binutils-2.20.1.orig/configure binutils-2.20.1/configure
--- binutils-2.20.1.orig/configure	2010-08-17 19:32:07.000000000 +0200
+++ binutils-2.20.1/configure	2010-08-17 19:32:22.000000000 +0200
@@ -3086,7 +3086,7 @@
     noconfigdirs="$noconfigdirs target-libffi target-qthreads"
     libgloss_dir=arm
     ;;
-  arm*-*-linux-gnueabi)
+  arm*-*-linux-*gnueabi)
     noconfigdirs="$noconfigdirs target-qthreads"
     case ${with_newlib} in
       no) noconfigdirs="$noconfigdirs target-newlib target-libgloss"
diff -durN binutils-2.20.1.orig/configure.ac binutils-2.20.1/configure.ac
--- binutils-2.20.1.orig/configure.ac	2010-08-17 19:32:07.000000000 +0200
+++ binutils-2.20.1/configure.ac	2010-08-17 19:32:22.000000000 +0200
@@ -573,7 +573,7 @@
     noconfigdirs="$noconfigdirs target-libffi target-qthreads"
     libgloss_dir=arm
     ;;
-  arm*-*-linux-gnueabi)
+  arm*-*-linux-*gnueabi)
     noconfigdirs="$noconfigdirs target-qthreads"
     case ${with_newlib} in
       no) noconfigdirs="$noconfigdirs target-newlib target-libgloss"
