--- binutils/ld/emulparams/elf_i386_glibc21.sh.jj	Sat Aug  3 17:04:41 2002
+++ binutils/ld/emulparams/elf_i386_glibc21.sh	Sat Aug  3 17:04:41 2002
@@ -0,0 +1,11 @@
+SCRIPT_NAME=elf
+OUTPUT_FORMAT="elf32-i386"
+TEXT_START_ADDR=0x08048000
+MAXPAGESIZE=0x1000
+NONPAGED_TEXT_START_ADDR=0x08048000
+ARCH=i386
+MACHINE=
+NOP=0x9090
+TEMPLATE_NAME=elf32
+GENERATE_SHLIB_SCRIPT=yes
+LIB_PATH=/usr/i386-glibc21-linux/lib:/usr/local/lib
--- binutils/ld/configure.tgt.jj	Thu Aug  1 20:49:32 2002
+++ binutils/ld/configure.tgt	Sat Aug  3 17:06:18 2002
@@ -143,7 +143,8 @@ i[3456]86-*-linux*aout*)	targ_emul=i386l
 			;;
 i[3456]86-*-linux*oldld)	targ_emul=i386linux; targ_extra_emuls=elf_i386 ;;
 i[3456]86-*-linux-gnu*)	targ_emul=elf_i386
-			targ_extra_emuls=i386linux
+			targ_extra_emuls="i386linux elf_i386_glibc21"
+			targ_extra_libpath=elf_i386_glibc21
 			if test x${want64} = xtrue; then
 			  targ_extra_emuls="$targ_extra_emuls elf_x86_64"
 			fi
--- binutils/ld/Makefile.am.jj	Thu Aug  1 20:49:32 2002
+++ binutils/ld/Makefile.am	Sat Aug  3 17:04:41 2002
@@ -181,6 +181,7 @@ ALL_EMULATIONS = \
 	eelf_i386.o \
 	eelf_i386_be.o \
 	eelf_i386_chaos.o \
+	eelf_i386_glibc21.o \
 	eelf_i386_fbsd.o \
 	eelf_i386_ldso.o \
 	eelf_s390.o \
@@ -718,6 +719,9 @@ eelf_i386_be.c: $(srcdir)/emulparams/elf
 eelf_i386_chaos.c: $(srcdir)/emulparams/elf_i386_chaos.sh \
   $(srcdir)/emultempl/elf32.em $(srcdir)/scripttempl/elf_chaos.sc ${GEN_DEPENDS}
 	${GENSCRIPTS} elf_i386_chaos "$(tdir_elf_i386_chaos)"
+eelf_i386_glibc21.c: $(srcdir)/emulparams/elf_i386_glibc21.sh \
+  $(srcdir)/emultempl/elf32.em $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
+	${GENSCRIPTS} elf_i386_glibc21 "$(tdir_elf_i386_glibc21)"
 eelf_i386_fbsd.c: $(srcdir)/emulparams/elf_i386_fbsd.sh \
   $(srcdir)/emulparams/elf_i386.sh \
   $(srcdir)/emultempl/elf32.em $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
--- binutils/ld/Makefile.in.jj	Thu Aug  1 20:49:32 2002
+++ binutils/ld/Makefile.in	Sat Aug  3 17:07:34 2002
@@ -1197,6 +1197,9 @@ eelf_i386_be.c: $(srcdir)/emulparams/elf
 eelf_i386_chaos.c: $(srcdir)/emulparams/elf_i386_chaos.sh \
   $(srcdir)/emultempl/elf32.em $(srcdir)/scripttempl/elf_chaos.sc ${GEN_DEPENDS}
 	${GENSCRIPTS} elf_i386_chaos "$(tdir_elf_i386_chaos)"
+eelf_i386_glibc21.c: $(srcdir)/emulparams/elf_i386_glibc21.sh \
+  $(srcdir)/emultempl/elf32.em $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
+	${GENSCRIPTS} elf_i386_glibc21 "$(tdir_elf_i386_glibc21)"
 eelf_i386_fbsd.c: $(srcdir)/emulparams/elf_i386_fbsd.sh \
   $(srcdir)/emulparams/elf_i386.sh \
   $(srcdir)/emultempl/elf32.em $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
