diff -urpN newlib-linaro-2.1.0-2014.09.orig/libgloss/arm/configure newlib-linaro-2.1.0-2014.09/libgloss/arm/configure
--- newlib-linaro-2.1.0-2014.09.orig/libgloss/arm/configure	2017-04-11 00:41:25.484880978 -0700
+++ newlib-linaro-2.1.0-2014.09/libgloss/arm/configure	2017-04-11 00:43:12.341789130 -0700
@@ -2531,7 +2531,7 @@ test "${CCASFLAGS+set}" = set || CCASFLA
 
 
 case "${target}" in
-  *-*-elf | *-*-eabi | *-*-tirtos*)
+  *-*-elf | *-*-eabi* | *-*-tirtos*)
 	objtype=elf-
 	;;
   *-*-coff)
diff -urpN newlib-linaro-2.1.0-2014.09.orig/libgloss/arm/configure.in newlib-linaro-2.1.0-2014.09/libgloss/arm/configure.in
--- newlib-linaro-2.1.0-2014.09.orig/libgloss/arm/configure.in	2017-04-11 00:41:25.484880978 -0700
+++ newlib-linaro-2.1.0-2014.09/libgloss/arm/configure.in	2017-04-11 00:43:12.341789130 -0700
@@ -49,7 +49,7 @@ AC_PROG_RANLIB
 LIB_AM_PROG_AS
 
 case "${target}" in
-  *-*-elf | *-*-eabi | *-*-tirtos*)
+  *-*-elf | *-*-eabi* | *-*-tirtos*)
 	objtype=elf-
 	;;
   *-*-coff)
