--- gcc-4.3.2.orig/gcc/config/arm/linux-eabi.h	2009-09-08 23:46:44.000000000 +0200
+++ gcc-4.3.2/gcc/config/arm/linux-eabi.h	2009-09-08 23:46:51.000000000 +0200
@@ -44,7 +44,7 @@
    The ARM10TDMI core is the default for armv5t, so set
    SUBTARGET_CPU_DEFAULT to achieve this.  */
 #undef  SUBTARGET_CPU_DEFAULT
-#define SUBTARGET_CPU_DEFAULT TARGET_CPU_arm10tdmi
+#define SUBTARGET_CPU_DEFAULT TARGET_CPU_arm9tdmi
 
 /* TARGET_BIG_ENDIAN_DEFAULT is set in
    config.gcc for big endian configurations.  */
