diff -durN gcc-4.5.2.orig/libjava/Makefile.in gcc-4.5.2/libjava/Makefile.in
--- gcc-4.5.2.orig/libjava/Makefile.in	2010-12-16 13:49:03.000000000 +0100
+++ gcc-4.5.2/libjava/Makefile.in	2010-12-29 23:02:41.000000000 +0100
@@ -9463,6 +9463,9 @@
 ecjx$(EXEEXT): $(ecjx_OBJECTS) $(ecjx_DEPENDENCIES) 
 	@rm -f ecjx$(EXEEXT)
 	$(ecjx_LINK) $(ecjx_OBJECTS) $(ecjx_LDADD) $(LIBS)
+ecjx.$(OBJEXT): $(ecjx_SOURCES)
+	@rm -f ecjx.$(OBJEXT)
+	$(CC_FOR_BUILD) $(BUILD_CFLAGS) -c -o $@ $<
 gappletviewer$(EXEEXT): $(gappletviewer_OBJECTS) $(gappletviewer_DEPENDENCIES) 
 	@rm -f gappletviewer$(EXEEXT)
 	$(gappletviewer_LINK) $(gappletviewer_OBJECTS) $(gappletviewer_LDADD) $(LIBS)
