--- ppl-0.10.2/configure	2010-03-19 07:44:26.000000000 +0100
+++ ppl-0.10.2-new/configure	2010-03-19 19:25:02.000000000 +0100
@@ -15603,11 +15603,12 @@
                 found_la="$additional_libdir/lib$name.la"
               fi
             else
-              if test -f "$additional_libdir/lib$name.$libext"; then
+              post=${shlibext:+.${shlibext}}
+              if test -f "$additional_libdir/lib$name$post.$libext"; then
                 found_dir="$additional_libdir"
-                found_a="$additional_libdir/lib$name.$libext"
-                if test -f "$additional_libdir/lib$name.la"; then
-                  found_la="$additional_libdir/lib$name.la"
+                found_a="$additional_libdir/lib$name$post.$libext"
+                if test -f "$additional_libdir/lib$name$post.la"; then
+                  found_la="$additional_libdir/lib$name$post.la"
                 fi
               fi
             fi
