diff -Naurd cloog-ppl-0.15.11-a/configure cloog-ppl-0.15.11-b/configure
--- cloog-ppl-0.15.11-a/configure	2011-04-01 06:22:53.000000000 +0200
+++ cloog-ppl-0.15.11-b/configure	2012-01-27 13:29:42.698987517 +0100
@@ -11238,7 +11238,7 @@
 fi
 rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
 
-	LIBS="$LIBS -lppl_c -lppl -lgmpxx"
+	LIBS="-lppl_c -lppl -lgmpxx $LIBS"
 
 $as_echo "#define CLOOG_PPL_BACKEND 1" >>confdefs.h
 
diff -Naurd cloog-ppl-0.15.11-a/configure.in cloog-ppl-0.15.11-b/configure.in
--- cloog-ppl-0.15.11-a/configure.in	2010-08-27 21:13:48.000000000 +0200
+++ cloog-ppl-0.15.11-b/configure.in	2012-01-27 13:29:42.698987517 +0100
@@ -343,7 +343,7 @@
 	#endif
 	], [AC_MSG_RESULT([yes])], [AC_MSG_ERROR(Can't find correct version of PPL.) ])
 
-	LIBS="$LIBS -lppl_c -lppl -lgmpxx"
+	LIBS="-lppl_c -lppl -lgmpxx $LIBS"
 	AC_DEFINE([CLOOG_PPL_BACKEND], 1, [Use the PPL backend])
 
 else
