diff -durN gcc-4.4.5.orig/libgomp/configure gcc-4.4.5/libgomp/configure
--- gcc-4.4.5.orig/libgomp/configure	2010-10-01 10:26:18.000000000 +0200
+++ gcc-4.4.5/libgomp/configure	2010-10-09 23:00:33.000000000 +0200
@@ -3334,7 +3334,7 @@
 
 # Add -Wall -Werror if we are using GCC.
 if test "x$GCC" = "xyes"; then
-  XCFLAGS="$XCFLAGS -Wall -Werror"
+  XCFLAGS="$XCFLAGS -Wall"
 fi
 
 # Find other programs we need.
