diff -durN gcc-4.4.5.orig/gcc/config/ia64/crtbegin.asm gcc-4.4.5/gcc/config/ia64/crtbegin.asm
--- gcc-4.4.5.orig/gcc/config/ia64/crtbegin.asm	2009-04-10 01:23:07.000000000 +0200
+++ gcc-4.4.5/gcc/config/ia64/crtbegin.asm	2010-10-09 23:03:55.000000000 +0200
@@ -252,3 +252,7 @@
 .weak __cxa_finalize
 #endif
 .weak _Jv_RegisterClasses
+
+#ifdef __linux__
+.section .note.GNU-stack; .previous
+#endif
diff -durN gcc-4.4.5.orig/gcc/config/ia64/crtend.asm gcc-4.4.5/gcc/config/ia64/crtend.asm
--- gcc-4.4.5.orig/gcc/config/ia64/crtend.asm	2009-04-10 01:23:07.000000000 +0200
+++ gcc-4.4.5/gcc/config/ia64/crtend.asm	2010-10-09 23:03:55.000000000 +0200
@@ -119,3 +119,7 @@
 
 	br.ret.sptk.many rp
 	.endp __do_global_ctors_aux
+
+#ifdef __linux__
+.section .note.GNU-stack; .previous
+#endif
diff -durN gcc-4.4.5.orig/gcc/config/ia64/crti.asm gcc-4.4.5/gcc/config/ia64/crti.asm
--- gcc-4.4.5.orig/gcc/config/ia64/crti.asm	2009-04-10 01:23:07.000000000 +0200
+++ gcc-4.4.5/gcc/config/ia64/crti.asm	2010-10-09 23:03:55.000000000 +0200
@@ -51,3 +51,7 @@
 	.body
 
 # end of crti.asm
+
+#ifdef __linux__
+.section .note.GNU-stack; .previous
+#endif
diff -durN gcc-4.4.5.orig/gcc/config/ia64/crtn.asm gcc-4.4.5/gcc/config/ia64/crtn.asm
--- gcc-4.4.5.orig/gcc/config/ia64/crtn.asm	2009-04-10 01:23:07.000000000 +0200
+++ gcc-4.4.5/gcc/config/ia64/crtn.asm	2010-10-09 23:03:55.000000000 +0200
@@ -41,3 +41,7 @@
 	br.ret.sptk.many b0
 
 # end of crtn.asm
+
+#ifdef __linux__
+.section .note.GNU-stack; .previous
+#endif
diff -durN gcc-4.4.5.orig/gcc/config/ia64/lib1funcs.asm gcc-4.4.5/gcc/config/ia64/lib1funcs.asm
--- gcc-4.4.5.orig/gcc/config/ia64/lib1funcs.asm	2009-04-10 01:23:07.000000000 +0200
+++ gcc-4.4.5/gcc/config/ia64/lib1funcs.asm	2010-10-09 23:03:55.000000000 +0200
@@ -793,3 +793,7 @@
 	.endp __floattitf
 #endif
 #endif
+
+#ifdef __linux__
+.section .note.GNU-stack; .previous
+#endif
diff -durN gcc-4.4.5.orig/gcc/config/ia64/linux.h gcc-4.4.5/gcc/config/ia64/linux.h
--- gcc-4.4.5.orig/gcc/config/ia64/linux.h	2009-02-12 17:30:53.000000000 +0100
+++ gcc-4.4.5/gcc/config/ia64/linux.h	2010-10-09 23:03:55.000000000 +0200
@@ -5,6 +5,8 @@
 
 #define TARGET_VERSION fprintf (stderr, " (IA-64) Linux");
 
+#define TARGET_ASM_FILE_END file_end_indicate_exec_stack
+
 /* This is for -profile to use -lc_p instead of -lc.  */
 #undef CC1_SPEC
 #define CC1_SPEC "%{profile:-p} %{G*}"
diff -durN gcc-4.4.5.orig/gcc/config/rs6000/ppc-asm.h gcc-4.4.5/gcc/config/rs6000/ppc-asm.h
--- gcc-4.4.5.orig/gcc/config/rs6000/ppc-asm.h	2008-10-13 17:44:26.000000000 +0200
+++ gcc-4.4.5/gcc/config/rs6000/ppc-asm.h	2010-10-09 23:03:55.000000000 +0200
@@ -172,7 +172,7 @@
 	.size FUNC_NAME(name),GLUE(.L,name)-FUNC_NAME(name)
 #endif
 
-#if defined __linux__ && !defined __powerpc64__
+#if defined __linux__
 	.section .note.GNU-stack
 	.previous
 #endif
