diff -durN gcc-4.4.5.orig/gcc/config/sh/lib1funcs.asm gcc-4.4.5/gcc/config/sh/lib1funcs.asm
--- gcc-4.4.5.orig/gcc/config/sh/lib1funcs.asm	2009-04-10 01:23:07.000000000 +0200
+++ gcc-4.4.5/gcc/config/sh/lib1funcs.asm	2010-10-09 23:04:22.000000000 +0200
@@ -30,6 +30,11 @@
 !! recoded in assembly by Toshiyasu Morita
 !! tm@netcom.com
 
+#if defined(__ELF__) && defined(__linux__)
+.section .note.GNU-stack,"",%progbits
+.previous
+#endif
+
 /* SH2 optimizations for ___ashrsi3, ___ashlsi3, ___lshrsi3 and
    ELF local label prefixes by J"orn Rennecke
    amylaar@cygnus.com  */
diff -durN gcc-4.4.5.orig/gcc/config/sh/linux-atomic.asm gcc-4.4.5/gcc/config/sh/linux-atomic.asm
--- gcc-4.4.5.orig/gcc/config/sh/linux-atomic.asm	2009-04-10 01:23:07.000000000 +0200
+++ gcc-4.4.5/gcc/config/sh/linux-atomic.asm	2010-10-09 23:04:22.000000000 +0200
@@ -136,3 +136,6 @@
 ATOMIC_FETCH_AND_COMBOP(nand,and,not,4,l,mov)
 
 #endif /* ! __SH5__ */
+
+.section .note.GNU-stack,"",%progbits
+.previous
