--- uClibc-0.9.30.orig/extra/scripts/install_headers.sh	2008-10-10 15:50:38.000000000 +0200
+++ uClibc-0.9.30/extra/scripts/install_headers.sh	2011-08-21 09:49:03.000000000 +0200
@@ -33,7 +33,7 @@
 (
 # We must cd, or else we'll prepend "$1" to filenames!
 cd "$1" || exit 1
-find ! -name '.' -a ! -path '*/.*' | sed -e 's/^\.\///' -e '/^config\//d' \
+find . ! -name '.' -a ! -path '*/.*' | sed -e 's/^\.\///' -e '/^config\//d' \
 	-e '/^config$/d'
 ) | \
 (
