diff -durN gcc-4.4.0.orig/gcc/configure gcc-4.4.0/gcc/configure
--- gcc-4.4.0.orig/gcc/configure	2009-03-24 18:46:03.000000000 +0100
+++ gcc-4.4.0/gcc/configure	2009-05-27 21:38:06.000000000 +0200
@@ -13997,7 +13997,7 @@
 	    | powerpc*-*-*,powerpc64*-*-*)
 		CROSS="$CROSS -DNATIVE_CROSS" ;;
 	esac
-elif test "x$TARGET_SYSTEM_ROOT" != x; then
+elif test "x$TARGET_SYSTEM_ROOT" != x -o $build != $host; then
         SYSTEM_HEADER_DIR=$build_system_header_dir
 fi
 
diff -durN gcc-4.4.0.orig/gcc/configure.ac gcc-4.4.0/gcc/configure.ac
--- gcc-4.4.0.orig/gcc/configure.ac	2009-03-24 18:46:03.000000000 +0100
+++ gcc-4.4.0/gcc/configure.ac	2009-05-27 21:38:06.000000000 +0200
@@ -1720,7 +1720,7 @@
 	    | powerpc*-*-*,powerpc64*-*-*)
 		CROSS="$CROSS -DNATIVE_CROSS" ;;
 	esac
-elif test "x$TARGET_SYSTEM_ROOT" != x; then
+elif test "x$TARGET_SYSTEM_ROOT" != x -o $build != $host; then
         SYSTEM_HEADER_DIR=$build_system_header_dir 
 fi
 
diff -durN gcc-4.4.0.orig/gcc/unwind-dw2.c gcc-4.4.0/gcc/unwind-dw2.c
--- gcc-4.4.0.orig/gcc/unwind-dw2.c	2009-04-10 01:23:07.000000000 +0200
+++ gcc-4.4.0/gcc/unwind-dw2.c	2009-05-27 21:38:06.000000000 +0200
@@ -329,9 +329,11 @@
 }
 #endif
 
+#ifndef inhibit_libc
 #ifdef MD_UNWIND_SUPPORT
 #include MD_UNWIND_SUPPORT
 #endif
+#endif
 
 /* Extract any interesting information from the CIE for the translation
    unit F belongs to.  Return a pointer to the byte after the augmentation,
