diff -dur binutils-2.17.old/ld/Makefile.am binutils-2.17/ld/Makefile.am
--- binutils-2.17.old/ld/Makefile.am	2006-06-03 06:45:50.000000000 +0200
+++ binutils-2.17/ld/Makefile.am	2007-05-01 18:26:09.000000000 +0200
@@ -20,7 +20,7 @@
 # We put the scripts in the directory $(scriptdir)/ldscripts.
 # We can't put the scripts in $(datadir) because the SEARCH_DIR
 # directives need to be different for native and cross linkers.
-scriptdir = $(tooldir)/lib
+scriptdir = $(libdir)
 
 EMUL = @EMUL@
 EMULATION_OFILES = @EMULATION_OFILES@
diff -dur binutils-2.17.old/ld/Makefile.in binutils-2.17/ld/Makefile.in
--- binutils-2.17.old/ld/Makefile.in	2006-06-03 06:45:50.000000000 +0200
+++ binutils-2.17/ld/Makefile.in	2007-05-01 18:26:09.000000000 +0200
@@ -269,7 +269,7 @@
 # We put the scripts in the directory $(scriptdir)/ldscripts.
 # We can't put the scripts in $(datadir) because the SEARCH_DIR
 # directives need to be different for native and cross linkers.
-scriptdir = $(tooldir)/lib
+scriptdir = $(libdir)
 BASEDIR = $(srcdir)/..
 BFDDIR = $(BASEDIR)/bfd
 INCDIR = $(BASEDIR)/include
