diff -ur gcc-4.3.4-orig/libjava/Makefile.in gcc-4.3.4/libjava/Makefile.in
--- gcc-4.3.4-orig/libjava/Makefile.in	2009-08-04 04:12:27.000000000 -0700
+++ gcc-4.3.4/libjava/Makefile.in	2009-09-23 22:03:15.750951483 -0700
@@ -8874,6 +8874,9 @@
 ecjx$(EXEEXT): $(ecjx_OBJECTS) $(ecjx_DEPENDENCIES) 
 	@rm -f ecjx$(EXEEXT)
 	$(ecjx_LINK) $(ecjx_LDFLAGS) $(ecjx_OBJECTS) $(ecjx_LDADD) $(LIBS)
+ecjx.$(OBJEXT): $(ecjx_SOURCES)
+	@rm -f ecjx.$(OBJEXT)
+	$(CC_FOR_BUILD) $(BUILD_CFLAGS) -c -o $@ $<
 gappletviewer$(EXEEXT): $(gappletviewer_OBJECTS) $(gappletviewer_DEPENDENCIES) 
 	@rm -f gappletviewer$(EXEEXT)
 	$(gappletviewer_LINK) $(gappletviewer_LDFLAGS) $(gappletviewer_OBJECTS) $(gappletviewer_LDADD) $(LIBS)
