diff -durN gcc-4.4.5.orig/gcc/builtins.c gcc-4.4.5/gcc/builtins.c
--- gcc-4.4.5.orig/gcc/builtins.c	2010-03-22 16:00:20.000000000 +0100
+++ gcc-4.4.5/gcc/builtins.c	2010-10-09 22:57:58.000000000 +0200
@@ -5783,6 +5783,9 @@
   trampolines_created = 1;
   INITIALIZE_TRAMPOLINE (r_tramp, r_func, r_chain);
 
+  if (warn_trampolines)
+    warning (OPT_Wtrampolines, "generating trampoline in object (requires executable stack)");
+
   return const0_rtx;
 }
 
diff -durN gcc-4.4.5.orig/gcc/common.opt gcc-4.4.5/gcc/common.opt
--- gcc-4.4.5.orig/gcc/common.opt	2009-03-28 18:28:45.000000000 +0100
+++ gcc-4.4.5/gcc/common.opt	2010-10-09 22:57:58.000000000 +0200
@@ -197,6 +197,10 @@
 Common Var(warn_type_limits) Init(-1) Warning
 Warn if a comparison is always true or always false due to the limited range of the data type
 
+Wtrampolines
+Common Var(warn_trampolines) Init(1)
+Warn whenever a trampoline is generated
+
 Wuninitialized
 Common Var(warn_uninitialized) Warning
 Warn about uninitialized automatic variables
