diff -durN gcc-4.4.5.orig/libiberty/Makefile.in gcc-4.4.5/libiberty/Makefile.in
--- gcc-4.4.5.orig/libiberty/Makefile.in	2008-10-22 15:30:19.000000000 +0200
+++ gcc-4.4.5/libiberty/Makefile.in	2010-10-09 23:02:28.000000000 +0200
@@ -227,6 +227,7 @@
 	  $(AR) $(AR_FLAGS) $(TARGETLIB) \
 	    $(REQUIRED_OFILES) $(EXTRA_OFILES) $(LIBOBJS); \
 	  $(RANLIB) $(TARGETLIB); \
+	  cp $(TARGETLIB) ../ ; \
 	  cd ..; \
 	else true; fi
 
