diff -durN gcc-4.4.0.orig/libgomp/configure gcc-4.4.0/libgomp/configure

diff -durN gcc-4.4.4.orig/libgomp/configure gcc-4.4.4/libgomp/configure
--- gcc-4.4.4.orig/libgomp/configure	2010-04-29 10:26:54.000000000 +0200
+++ gcc-4.4.4/libgomp/configure	2010-05-16 19:12:19.000000000 +0200
@@ -3334,7 +3334,7 @@
 
 # Add -Wall -Werror if we are using GCC.
 if test "x$GCC" = "xyes"; then
-  XCFLAGS="$XCFLAGS -Wall -Werror"
+  XCFLAGS="$XCFLAGS -Wall"
 fi
 
 # Find other programs we need.
