diff -durN gcc-4.4.0.orig/libiberty/Makefile.in gcc-4.4.0/libiberty/Makefile.in

diff -durN gcc-4.4.4.orig/libiberty/Makefile.in gcc-4.4.4/libiberty/Makefile.in
--- gcc-4.4.4.orig/libiberty/Makefile.in	2008-10-22 15:30:19.000000000 +0200
+++ gcc-4.4.4/libiberty/Makefile.in	2010-05-16 19:13:20.000000000 +0200
@@ -227,6 +227,7 @@
 	  $(AR) $(AR_FLAGS) $(TARGETLIB) \
 	    $(REQUIRED_OFILES) $(EXTRA_OFILES) $(LIBOBJS); \
 	  $(RANLIB) $(TARGETLIB); \
+	  cp $(TARGETLIB) ../ ; \
 	  cd ..; \
 	else true; fi
 
