diff -durN gcc-4.4.4.orig/gcc/configure gcc-4.4.4/gcc/configure
--- gcc-4.4.4.orig/gcc/configure	2010-05-16 19:11:00.000000000 +0200
+++ gcc-4.4.4/gcc/configure	2010-05-16 19:14:31.000000000 +0200
@@ -22241,7 +22241,7 @@
 	tls_first_minor=14
 	tls_as_opt="-m64 -Aesame --fatal-warnings"
 	;;
-  sh-*-* | sh[34]-*-*)
+  sh-*-* | sh[34]*-*-*)
     conftest_s='
 	.section ".tdata","awT",@progbits
 foo:	.long	25
diff -durN gcc-4.4.4.orig/gcc/configure.ac gcc-4.4.4/gcc/configure.ac
--- gcc-4.4.4.orig/gcc/configure.ac	2010-05-16 19:11:00.000000000 +0200
+++ gcc-4.4.4/gcc/configure.ac	2010-05-16 19:14:31.000000000 +0200
@@ -2684,7 +2684,7 @@
 	tls_first_minor=14
 	tls_as_opt="-m64 -Aesame --fatal-warnings"
 	;;
-  sh-*-* | sh[34]-*-*)
+  sh-*-* | sh[34]*-*-*)
     conftest_s='
 	.section ".tdata","awT",@progbits
 foo:	.long	25
