commit 25bb77587944768581225d1c0048facab32cf019
Author: Roberto Bagnara <bagnara@cs.unipr.it>
Date:   Wed May 18 09:35:36 2011 +0200

    Moved misplaced conditional.

diff --git a/interfaces/Java/tests/Makefile.am b/interfaces/Java/tests/Makefile.am
index 64ba864..607bbc6 100644
--- a/interfaces/Java/tests/Makefile.am
+++ b/interfaces/Java/tests/Makefile.am
@@ -51,6 +51,8 @@ Parma_Polyhedra_Library_test2.java \
 PPL_Test.java \
 ppl_java_tests_common
 
+if ENABLE_SHARED
+
 check_JAVA = \
 ppl_java_generated_tests.java \
 Test_Executor.java \
@@ -62,8 +64,6 @@ Parma_Polyhedra_Library_test1.java \
 Parma_Polyhedra_Library_test2.java \
 PPL_Test.java
 
-if ENABLE_SHARED
-
 CLASSPATH = ../ppl_java.jar:.
 JAVACFLAGS = -classpath $(CLASSPATH)
 
