--- trunk/src/acos.c	2009/12/08 16:49:43	733
+++ trunk/src/acos.c	2009/12/09 20:51:38	734
@@ -188,10 +188,7 @@
   else
     rnd_im = rnd_im == GMP_RNDU ? GMP_RNDD
       : rnd_im == GMP_RNDD ? GMP_RNDU
-#if MPFR_VERSION_MAJOR >= 3
-      : rnd_im == GMP_RNDA ? GMP_RNDZ
-#endif
-      : rnd_im;
+      : rnd_im; /* both RNDZ and RNDA map to themselves for -asin(z) */
   rnd1 = RNDC(GMP_RNDN, rnd_im);
   mpfr_init2 (pi_over_2, p);
   for (;;)
