diff -durN binutils-2.14.orig/gas/config/tc-i386.c binutils-2.14/gas/config/tc-i386.c
--- binutils-2.14.orig/gas/config/tc-i386.c	2003-01-29 11:05:52.000000000 +0100
+++ binutils-2.14/gas/config/tc-i386.c	2008-09-22 08:54:50.000000000 +0200
@@ -189,15 +189,10 @@
 const char extra_symbol_chars[] = "*%-([";
 #endif
 
-#if (defined (TE_I386AIX)				\
-     || ((defined (OBJ_ELF) || defined (OBJ_MAYBE_ELF))	\
-	 && !defined (TE_LINUX)				\
-	 && !defined (TE_FreeBSD)			\
-	 && !defined (TE_NetBSD)))
 /* This array holds the chars that always start a comment.  If the
    pre-processor is disabled, these aren't very useful.  */
-const char comment_chars[] = "#/";
-#define PREFIX_SEPARATOR '\\'
+const char comment_chars[] = "#";
+#define PREFIX_SEPARATOR '/'
 
 /* This array holds the chars that only start a comment at the beginning of
    a line.  If the line seems to have the form '# 123 filename'
@@ -207,16 +202,7 @@
    #NO_APP at the beginning of its output.
    Also note that comments started like this one will always work if
    '/' isn't otherwise defined.  */
-const char line_comment_chars[] = "#";
-
-#else
-/* Putting '/' here makes it impossible to use the divide operator.
-   However, we need it for compatibility with SVR4 systems.  */
-const char comment_chars[] = "#";
-#define PREFIX_SEPARATOR '/'
-
 const char line_comment_chars[] = "/#";
-#endif
 
 const char line_separator_chars[] = ";";
 
