diff -ru eglibc-2_9_orig/elf/Makefile eglibc-2_9/elf/Makefile
--- eglibc-2_9_orig/elf/Makefile	2009-01-26 07:31:40.000000000 -0800
+++ eglibc-2_9/elf/Makefile	2009-05-06 16:52:28.000000000 -0700
@@ -305,7 +305,7 @@
 		  $(LDFLAGS-rtld) -Wl,-z,defs -Wl,--verbose 2>&1 |	\
 		  LC_ALL=C \
 		  sed -e '/^=========/,/^=========/!d;/^=========/d'	\
-		      -e 's/\. = 0 + SIZEOF_HEADERS;/& _begin = . - SIZEOF_HEADERS;/' \
+		      -e 's/\. = .* + SIZEOF_HEADERS;/& _begin = . - SIZEOF_HEADERS;/' \
 		      -e 's/\. = SEGMENT_START("text-segment", 0) + SIZEOF_HEADERS;/& _begin = . - SIZEOF_HEADERS;/' \
 		  > $@.lds
 	$(LINK.o) -nostdlib -nostartfiles -shared -o $@			\
