diff -durN duma_2_5_12.orig/Makefile duma_2_5_12/Makefile
--- duma_2_5_12.orig/Makefile	2007-08-17 09:34:25.000000000 +0200
+++ duma_2_5_12/Makefile	2007-09-07 21:50:04.000000000 +0200
@@ -93,9 +93,6 @@
 # also define 'WIN32'

 # some defaults:
-CC=gcc
-CXX=g++
-AR=ar
 INSTALL=install
 RM=rm
 RMFORCE=rm -f
@@ -299,7 +296,7 @@

 createconf$(EXEPOSTFIX): createconf.o
 	- $(RMFORCE) createconf$(EXEPOSTFIX)
-	$(CC) $(CFLAGS) $(DUMA_OPTIONS) createconf.o -o createconf$(EXEPOSTFIX)
+	$(HOSTCC) $(HOSTCFLAGS) $(DUMA_OPTIONS) createconf.o -o createconf$(EXEPOSTFIX)

 tstheap$(EXEPOSTFIX): libduma.a tstheap.o
 	- $(RMFORCE) tstheap$(EXEPOSTFIX)
@@ -360,7 +357,7 @@
 # define rules how to build objects for createconf
 #
 createconf.o:
-	$(CC) $(CFLAGS) $(DUMA_OPTIONS) -c createconf.c -o $@
+	$(HOSTCC) $(HOSTCFLAGS) $(DUMA_OPTIONS) -c createconf.c -o $@


 #
