diff -durN duma_2_5_12.orig/Makefile duma_2_5_12/Makefile
--- duma_2_5_12.orig/Makefile	2007-09-07 21:57:45.000000000 +0200
+++ duma_2_5_12/Makefile	2007-09-07 21:58:35.000000000 +0200
@@ -231,9 +231,13 @@
 	dumatest.c tstheap.c thread-test.c testmt.c dumatestpp.cpp testoperators.cpp \
 	createconf.c

-OBJECTS = dumapp.o duma.o sem_inc.o print.o
+OBJECTS = duma.o sem_inc.o print.o
+SO_OBJECTS = duma_so.o sem_inc_so.o print_so.o

-SO_OBJECTS = dumapp_so.o duma_so.o sem_inc_so.o print_so.o
+ifeq ($(DUMA_CPP),1)
+OBJECTS += dumapp.o
+SO_OBJECTS += dumapp_so.o
+endif

 all:	libduma.a tstheap$(EXEPOSTFIX) dumatest$(EXEPOSTFIX) thread-test$(EXEPOSTFIX) testmt$(EXEPOSTFIX) dumatestpp$(EXEPOSTFIX) testoperators$(EXEPOSTFIX) $(DUMA_DYN_DEPS)
 	@ $(ECHOLF)
