diff -dur binutils-2.17.50.0.14.orig/bfd/configure binutils-2.17.50.0.14/bfd/configure
--- binutils-2.17.50.0.14.orig/bfd/configure	2007-03-23 16:41:48.000000000 +0100
+++ binutils-2.17.50.0.14/bfd/configure	2007-04-10 17:35:41.000000000 +0200
@@ -3577,7 +3577,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux-gnu*|linux-uclibc*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
diff -dur binutils-2.17.50.0.14.orig/binutils/configure binutils-2.17.50.0.14/binutils/configure
--- binutils-2.17.50.0.14.orig/binutils/configure	2007-03-16 16:48:29.000000000 +0100
+++ binutils-2.17.50.0.14/binutils/configure	2007-04-10 17:35:41.000000000 +0200
@@ -3410,7 +3410,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux-gnu*|linux-uclibc*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
diff -dur binutils-2.17.50.0.14.orig/configure binutils-2.17.50.0.14/configure
--- binutils-2.17.50.0.14.orig/configure	2007-03-23 16:41:48.000000000 +0100
+++ binutils-2.17.50.0.14/configure	2007-04-10 17:35:41.000000000 +0200
@@ -2195,7 +2195,7 @@
   am33_2.0-*-linux*)
     noconfigdirs="$noconfigdirs ${libgcj} target-newlib target-libgloss"
     ;;
-  sh-*-linux*)
+  sh*-*-linux*)
     noconfigdirs="$noconfigdirs ${libgcj} target-newlib target-libgloss"
     ;;
   sh*-*-pe|mips*-*-pe|*arm-wince-pe)
@@ -2490,7 +2490,7 @@
   romp-*-*)
     noconfigdirs="$noconfigdirs bfd binutils ld gas opcodes target-libgloss ${libgcj}"
     ;;
-  sh-*-* | sh64-*-*)
+  sh*-*-* | sh64-*-*)
     case "${host}" in
       i[3456789]86-*-vsta) ;; # don't add gprof back in
       i[3456789]86-*-go32*) ;; # don't add gprof back in
diff -dur binutils-2.17.50.0.14.orig/configure.ac binutils-2.17.50.0.14/configure.ac
--- binutils-2.17.50.0.14.orig/configure.ac	2007-03-23 16:41:48.000000000 +0100
+++ binutils-2.17.50.0.14/configure.ac	2007-04-10 17:35:41.000000000 +0200
@@ -484,7 +484,7 @@
   am33_2.0-*-linux*)
     noconfigdirs="$noconfigdirs ${libgcj} target-newlib target-libgloss"
     ;;
-  sh-*-linux*)
+  sh*-*-linux*)
     noconfigdirs="$noconfigdirs ${libgcj} target-newlib target-libgloss"
     ;;    
   sh*-*-pe|mips*-*-pe|*arm-wince-pe)
@@ -779,7 +779,7 @@
   romp-*-*)
     noconfigdirs="$noconfigdirs bfd binutils ld gas opcodes target-libgloss ${libgcj}"
     ;;
-  sh-*-* | sh64-*-*)
+  sh*-*-* | sh64-*-*)
     case "${host}" in
       i[[3456789]]86-*-vsta) ;; # don't add gprof back in
       i[[3456789]]86-*-go32*) ;; # don't add gprof back in
diff -dur binutils-2.17.50.0.14.orig/gas/configure binutils-2.17.50.0.14/gas/configure
--- binutils-2.17.50.0.14.orig/gas/configure	2007-03-16 16:48:29.000000000 +0100
+++ binutils-2.17.50.0.14/gas/configure	2007-04-10 17:35:41.000000000 +0200
@@ -3410,7 +3410,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux-gnu*|linux-uclibc*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
diff -dur binutils-2.17.50.0.14.orig/gprof/configure binutils-2.17.50.0.14/gprof/configure
--- binutils-2.17.50.0.14.orig/gprof/configure	2007-01-03 20:49:13.000000000 +0100
+++ binutils-2.17.50.0.14/gprof/configure	2007-04-10 17:35:41.000000000 +0200
@@ -3413,6 +3413,11 @@
   lt_cv_deplibs_check_method=pass_all
   ;;
 
+linux-uclibc*)
+  lt_cv_deplibs_check_method=pass_all
+  lt_cv_file_magic_test_file=`echo /lib/libuClibc-*.so`
+  ;;
+
 netbsd* | knetbsd*-gnu)
   if echo __ELF__ | $CC -E - | grep __ELF__ > /dev/null; then
     lt_cv_deplibs_check_method='match_pattern /lib[^/\.]+\.so\.[0-9]+\.[0-9]+$'
diff -dur binutils-2.17.50.0.14.orig/ld/configure binutils-2.17.50.0.14/ld/configure
--- binutils-2.17.50.0.14.orig/ld/configure	2007-03-16 16:48:30.000000000 +0100
+++ binutils-2.17.50.0.14/ld/configure	2007-04-10 17:35:41.000000000 +0200
@@ -3412,7 +3412,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux-gnu*|linux-uclibc*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
diff -dur binutils-2.17.50.0.14.orig/libtool.m4 binutils-2.17.50.0.14/libtool.m4
--- binutils-2.17.50.0.14.orig/libtool.m4	2006-02-14 00:10:15.000000000 +0100
+++ binutils-2.17.50.0.14/libtool.m4	2007-04-10 17:35:41.000000000 +0200
@@ -739,7 +739,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux-gnu*|linux-uclibc*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
diff -dur binutils-2.17.50.0.14.orig/ltconfig binutils-2.17.50.0.14/ltconfig
--- binutils-2.17.50.0.14.orig/ltconfig	2006-07-15 20:23:48.000000000 +0200
+++ binutils-2.17.50.0.14/ltconfig	2007-04-10 17:35:41.000000000 +0200
@@ -602,6 +602,7 @@
 
 # Transform linux* to *-*-linux-gnu*, to support old configure scripts.
 case $host_os in
+linux-uclibc*) ;;
 linux-gnu*) ;;
 linux*) host=`echo $host | sed 's/^\(.*-.*-linux\)\(.*\)$/\1-gnu\2/'`
 esac
@@ -1247,7 +1248,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux-gnu*|linux-uclibc*)
   version_type=linux
   need_lib_prefix=no
   need_version=no
diff -dur binutils-2.17.50.0.14.orig/opcodes/configure binutils-2.17.50.0.14/opcodes/configure
--- binutils-2.17.50.0.14.orig/opcodes/configure	2007-03-16 16:48:30.000000000 +0100
+++ binutils-2.17.50.0.14/opcodes/configure	2007-04-10 17:35:41.000000000 +0200
@@ -3578,7 +3578,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux-gnu*|linux-uclibc*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
