diff -dur binutils-2.17.50.0.14.orig/configure binutils-2.17.50.0.14/configure
--- binutils-2.17.50.0.14.orig/configure	2007-04-10 17:38:23.000000000 +0200
+++ binutils-2.17.50.0.14/configure	2007-04-10 17:38:43.000000000 +0200
@@ -2224,7 +2224,7 @@
   arm-*-elf* | strongarm-*-elf* | xscale-*-elf* | arm*-*-eabi* )
     noconfigdirs="$noconfigdirs target-libffi target-qthreads"
     ;;
-  arm*-*-linux-gnueabi)
+  arm*-*-linux-gnueabi | arm*-*-linux-uclibcgnueabi)
     noconfigdirs="$noconfigdirs target-libffi target-qthreads"
     noconfigdirs="$noconfigdirs target-libjava target-libobjc"
     ;;
diff -dur binutils-2.17.50.0.14.orig/configure.ac binutils-2.17.50.0.14/configure.ac
--- binutils-2.17.50.0.14.orig/configure.ac	2007-04-10 17:38:23.000000000 +0200
+++ binutils-2.17.50.0.14/configure.ac	2007-04-10 17:38:43.000000000 +0200
@@ -513,7 +513,7 @@
   arm-*-elf* | strongarm-*-elf* | xscale-*-elf* | arm*-*-eabi* )
     noconfigdirs="$noconfigdirs target-libffi target-qthreads"
     ;;
-  arm*-*-linux-gnueabi)
+  arm*-*-linux-gnueabi | arm*-*-linux-uclibcgnueabi)
     noconfigdirs="$noconfigdirs target-libffi target-qthreads"
     noconfigdirs="$noconfigdirs target-libjava target-libobjc"
     ;;
