diff -dur uClibc-0.9.28.1.orig/libc/sysdeps/linux/mips/clone.S uClibc-0.9.28.1/libc/sysdeps/linux/mips/clone.S
--- uClibc-0.9.28.1.orig/libc/sysdeps/linux/mips/clone.S	2007-02-07 11:18:12.000000000 +0100
+++ uClibc-0.9.28.1/libc/sysdeps/linux/mips/clone.S	2007-02-07 10:41:34.000000000 +0100
@@ -24,7 +24,7 @@
 #include <sys/regdef.h>
 #define _ERRNO_H	1
 #include <bits/errno.h>
-#include <asm/asm.h>
+#include <sys/asm.h>
 
 /* int clone(int (*fn)(void *arg), void *child_stack, int flags, void *arg) */
 
diff -dur uClibc-0.9.28.1.orig/libc/sysdeps/linux/mips/pipe.S uClibc-0.9.28.1/libc/sysdeps/linux/mips/pipe.S
--- uClibc-0.9.28.1.orig/libc/sysdeps/linux/mips/pipe.S	2007-02-07 11:18:12.000000000 +0100
+++ uClibc-0.9.28.1/libc/sysdeps/linux/mips/pipe.S	2007-02-07 11:16:52.000000000 +0100
@@ -3,9 +3,9 @@
 /*see uClibc's sh/pipe.c and glibc-2.2.4's mips/pipe.S */
 
 #include <features.h>
-#include <asm/asm.h>
+#include <sys/asm.h>
 #include <asm/unistd.h>
-#include <asm/regdef.h>
+#include <sys/regdef.h>
 
         .globl  pipe
         .ent    pipe, 0
diff -dur uClibc-0.9.28.1.orig/libc/sysdeps/linux/mips/syscall.S uClibc-0.9.28.1/libc/sysdeps/linux/mips/syscall.S
--- uClibc-0.9.28.1.orig/libc/sysdeps/linux/mips/syscall.S	2007-02-07 11:18:12.000000000 +0100
+++ uClibc-0.9.28.1/libc/sysdeps/linux/mips/syscall.S	2007-02-07 11:17:03.000000000 +0100
@@ -17,8 +17,8 @@
    02111-1307 USA.  */
 
 #include <features.h>
-#include <asm/asm.h>
-#include <asm/regdef.h>
+#include <sys/asm.h>
+#include <sys/regdef.h>
 
 #ifdef __PIC__
 	.option pic2
