diff -dur uClibc-0.9.28.orig/Rules.mak uClibc-0.9.28/Rules.mak
--- uClibc-0.9.28.orig/Rules.mak	2005-08-18 00:49:49.000000000 +0200
+++ uClibc-0.9.28/Rules.mak	2007-01-30 10:53:34.000000000 +0100
@@ -168,6 +168,7 @@
 	CPU_CFLAGS-$(CONFIG_MIPS_ISA_4)+=-mips4 -mtune=mips4
 	CPU_CFLAGS-$(CONFIG_MIPS_ISA_MIPS32)+=-mips32 -mtune=mips32
 	CPU_CFLAGS-$(CONFIG_MIPS_ISA_MIPS64)+=-mips64 -mtune=mips32
+	CPU_CFLAGS-$(CONFIG_MIPS_ISA_CUSTOM)+=-march=$(CONFIG_MIPS_CPU_CFLAGS_CUSTOM)
 endif
 
 ifeq ($(strip $(TARGET_ARCH)),sh)
diff -dur uClibc-0.9.28.orig/extra/Configs/Config.mips uClibc-0.9.28/extra/Configs/Config.mips
--- uClibc-0.9.28.orig/extra/Configs/Config.mips	2005-08-18 00:49:41.000000000 +0200
+++ uClibc-0.9.28/extra/Configs/Config.mips	2007-01-30 10:52:12.000000000 +0100
@@ -63,4 +63,16 @@
 config CONFIG_MIPS_ISA_MIPS64
 	bool "MIPS64"
 
+config CONFIG_MIPS_ISA_CUSTOM
+	bool "Custom"
+
 endchoice
+
+config CONFIG_MIPS_CPU_CFLAGS_CUSTOM
+	string
+	prompt "Custon ISA"
+	depends on CONFIG_MIPS_ISA_CUSTOM
+	default ""
+	help
+	  Enter your custom ISA here (eg: lx4189!).
+
