diff -dur strace-4.5.16.orig/system.c strace-4.5.16/system.c
--- strace-4.5.16.orig/system.c	2007-01-17 00:22:36.000000000 +0100
+++ strace-4.5.16/system.c	2007-07-14 19:28:41.000000000 +0200
@@ -1612,7 +1612,9 @@
 	{ CTL_KERN, "CTL_KERN" },
 	{ CTL_VM, "CTL_VM" },
 	{ CTL_NET, "CTL_NET" },
+#ifdef CTL_PROC
 	{ CTL_PROC, "CTL_PROC" },
+#endif
 	{ CTL_FS, "CTL_FS" },
 	{ CTL_DEBUG, "CTL_DEBUG" },
 	{ CTL_DEV, "CTL_DEV" },
