diff -ruN gettext-0.20.1.orig/gettext-tools/gnulib-tests/pthread_sigmask.c gettext-0.20.1/gettext-tools/gnulib-tests/pthread_sigmask.c
--- gettext-0.20.1.orig/gettext-tools/gnulib-tests/pthread_sigmask.c	2019-01-06 15:17:10.000000000 +0300
+++ gettext-0.20.1/gettext-tools/gnulib-tests/pthread_sigmask.c	2020-01-29 23:43:14.192282350 +0300
@@ -30,9 +30,9 @@
 # include <unistd.h>
 #endif
 
+#undef pthread_sigmask
 int
 pthread_sigmask (int how, const sigset_t *new_mask, sigset_t *old_mask)
-#undef pthread_sigmask
 {
 #if HAVE_PTHREAD_SIGMASK
   int ret;
