From 9b74f600cb69cf86c9a8e29e8cea775effd780c2 Mon Sep 17 00:00:00 2001
From: Chris Packham <chris.packham@alliedtelesis.co.nz>
Date: Sun, 25 May 2025 10:34:01 +1200
Subject: [PATCH] 0003-libgcc-disable-split-stack-nothreads.patch

---
 libgcc/config/t-stack | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/libgcc/config/t-stack b/libgcc/config/t-stack
index cc0366b4cd81..f3f97e86d60e 100644
--- a/libgcc/config/t-stack
+++ b/libgcc/config/t-stack
@@ -1,4 +1,6 @@
 # Makefile fragment to provide generic support for -fsplit-stack.
 # This should be used in config.host for any host which supports
 # -fsplit-stack.
+ifeq ($(enable_threads),yes)
 LIB2ADD_ST += $(srcdir)/generic-morestack.c $(srcdir)/generic-morestack-thread.c
+endif
-- 
2.49.0

