--- glibc-2.13/sysdeps/unix/sysv/linux/kernel-features.h	2010-07-27 14:34:39.000000000 +0300
+++ glibc-2.13/sysdeps/unix/sysv/linux/kernel-features.h.new	2011-03-10 18:54:37.686795979 +0200
@@ -437,7 +437,8 @@
    the code.  On PPC they were introduced in 2.6.17-rc1,
    on SH in 2.6.19-rc1.  */
 #if __LINUX_KERNEL_VERSION >= 0x020611 \
-    && (!defined __sh__ || __LINUX_KERNEL_VERSION >= 0x020613)
+    && (!defined __sh__ || __LINUX_KERNEL_VERSION >= 0x020613) \
+    && (!defined __alpha__)
 # define __ASSUME_ATFCTS	1
 #endif
 
