diff -ur dmalloc-5.5.2.orig/return.h dmalloc-5.5.2/return.h
--- dmalloc-5.5.2.orig/return.h	2016-03-13 13:11:48.090431764 -0700
+++ dmalloc-5.5.2/return.h	2016-03-13 13:12:11.246642618 -0700
@@ -251,8 +251,7 @@
 
 #define GET_RET_ADDR(file) \
 do { \
-  asm("mflr 0"); \
-  asm("stw 0,%0" : "=g" (file)); \
+  asm("mflr %0" : "=r"(file)); \
 } while(0)
 
 #endif /* __powerpc__ && __GNUC__ && !__OPTIMIZE__ */
