--- dmalloc-5.5.2.orig/configure	2008-06-17 13:10:09.000000000 +0200
+++ dmalloc-5.5.2/configure	2008-06-17 13:11:25.000000000 +0200
@@ -4691,8 +4691,35 @@
 echo "$as_me:$LINENO: checking strdup macro" >&5
 echo $ECHO_N "checking strdup macro... $ECHO_C" >&6
 if test "$cross_compiling" = yes; then
-  ac_cv_strdup_macro=no
+  cat >conftest.$ac_ext <<_ACEOF
+/* confdefs.h */
+_ACEOF
+cat confdefs.h >>conftest.$ac_ext
+cat >>conftest.$ac_ext <<_ACEOF
+/* end confdefs.h.  */
+
+#if HAVE_STDLIB_H
+#  include <string.h>
+#endif
+
+#ifndef strdup
+       choke me
+#endif
 
+main() { exit(0); }
+
+_ACEOF
+rm -f conftest.o conftest.obj
+if { (eval echo "$as_me:$LINENO: \"$ac_compile\"") >&5
+  (eval $ac_compile) 2>&5
+  ac_status=$?
+  echo "$as_me:$LINENO: \$? = $ac_status" >&5
+  (exit $ac_status); }; then
+  ac_cv_strdup_macro=yes
+else
+  ac_cv_strdup_macro=no
+fi
+  
 else
   cat >conftest.$ac_ext <<_ACEOF
 /* confdefs.h.  */
