--- mingw-libgnurx-2.5.1.orig/configure	2009-11-03 08:19:15.000000000 +0100
+++ mingw-libgnurx-2.5.1/configure	2009-11-03 08:31:37.000000000 +0100
@@ -273,7 +273,7 @@
 PACKAGE_STRING='MinGW libgnurx 2.5.1'
 PACKAGE_BUGREPORT='https://sourceforge.net/tracker/?group_id=2435&atid=102435'

-ac_subst_vars='SHELL PATH_SEPARATOR PACKAGE_NAME PACKAGE_TARNAME PACKAGE_VERSION PACKAGE_STRING PACKAGE_BUGREPORT exec_prefix prefix program_transform_name bindir sbindir libexecdir datadir sysconfdir sharedstatedir localstatedir libdir includedir oldincludedir infodir mandir build_alias host_alias target_alias DEFS ECHO_C ECHO_N ECHO_T LIBS CC CFLAGS LDFLAGS CPPFLAGS ac_ct_CC EXEEXT OBJEXT CC_QUALIFIED DLLVERSION ZIPCMD ZIPEXT MSVCLIB ac_ct_MSVCLIB GNURX_LIB install_dev LIBOBJS LTLIBOBJS'
+ac_subst_vars='SHELL PATH_SEPARATOR PACKAGE_NAME PACKAGE_TARNAME PACKAGE_VERSION PACKAGE_STRING PACKAGE_BUGREPORT exec_prefix prefix program_transform_name bindir sbindir libexecdir datadir sysconfdir sharedstatedir localstatedir libdir includedir oldincludedir infodir mandir build_alias host_alias target_alias DEFS ECHO_C ECHO_N ECHO_T LIBS CC CFLAGS LDFLAGS CPPFLAGS ac_ct_CC EXEEXT OBJEXT CC_QUALIFIED AR DLLVERSION ZIPCMD ZIPEXT MSVCLIB ac_ct_MSVCLIB GNURX_LIB install_dev LIBOBJS LTLIBOBJS'
 ac_subst_files=''

 # Initialize some variables set by options.
@@ -2190,21 +2190,54 @@

 fi
 rm -f conftest.err conftest.$ac_objext conftest.$ac_ext
-ac_ext=c
-ac_cpp='$CPP $CPPFLAGS'
-ac_compile='$CC -c $CFLAGS $CPPFLAGS conftest.$ac_ext >&5'
-ac_link='$CC -o conftest$ac_exeext $CFLAGS $CPPFLAGS $LDFLAGS conftest.$ac_ext $LIBS >&5'
-ac_compiler_gnu=$ac_cv_c_compiler_gnu

-  ac_ext=c
+
+for ac_prog in ar; do
+set dummy ${ac_tool_prefix}$ac_prog; ac_word=$2
+echo "$as_me:$LINENO: checking for $ac_word" >&5
+echo $ECHO_N "checking for $ac_word... $ECHO_C" >&6
+if test "${ac_cv_prog_AR+set}" = set; then
+  echo $ECHO_N "(cached) $ECHO_C" >&6
+else
+  if test -n "$AR"; then
+  ac_cv_prog_AR="$AR" # Let the user override the test.
+else
+as_save_IFS=$IFS; IFS=$PATH_SEPARATOR
+for as_dir in $PATH
+do
+  IFS=$as_save_IFS
+  test -z "$as_dir" && as_dir=.
+  for ac_exec_ext in '' $ac_executable_extensions; do
+  if $as_executable_p "$as_dir/$ac_word$ac_exec_ext"; then
+    ac_cv_prog_AR="$ac_tool_prefix$ac_prog"
+    echo "$as_me:$LINENO: found $as_dir/$ac_word$ac_exec_ext" >&5
+    break 2
+  fi
+done
+done
+
+  test -z "$ac_cv_prog_AR" && ac_cv_prog_AR="ar"
+fi
+fi
+AR=$ac_cv_prog_AR
+if test -n "$AR"; then
+  echo "$as_me:$LINENO: result: $AR" >&5
+echo "${ECHO_T}$AR" >&6
+else
+  echo "$as_me:$LINENO: result: no" >&5
+echo "${ECHO_T}no" >&6
+fi
+done
+
+ac_ext=c
 ac_cpp='$CPP $CPPFLAGS'
 ac_compile='$CC -c $CFLAGS $CPPFLAGS conftest.$ac_ext >&5'
 ac_link='$CC -o conftest$ac_exeext $CFLAGS $CPPFLAGS $LDFLAGS conftest.$ac_ext $LIBS >&5'
 ac_compiler_gnu=$ac_cv_c_compiler_gnu

- popCFLAGS=$CFLAGS
- echo 'int main(void){return 0;}' > conftest.$ac_ext
- echo "$as_me:$LINENO: checking whether $CC accepts the -mthreads option" >&5
+popCFLAGS=$CFLAGS
+echo 'int main(void){return 0;}' > conftest.$ac_ext
+echo "$as_me:$LINENO: checking whether $CC accepts the -mthreads option" >&5
 echo $ECHO_N "checking whether $CC accepts the -mthreads option... $ECHO_C" >&6
    CFLAGS="$popCFLAGS $CC_QUALIFIED -mthreads"
    if (eval $ac_compile) 2>&5; then
@@ -3038,6 +3069,7 @@
 s,@EXEEXT@,$EXEEXT,;t t
 s,@OBJEXT@,$OBJEXT,;t t
 s,@CC_QUALIFIED@,$CC_QUALIFIED,;t t
+s,@AR@,$AR,;t t
 s,@DLLVERSION@,$DLLVERSION,;t t
 s,@ZIPCMD@,$ZIPCMD,;t t
 s,@ZIPEXT@,$ZIPEXT,;t t
--- mingw-libgnurx-2.5.1.orig/Makefile.in	2009-11-03 08:17:49.000000000 +0100
+++ mingw-libgnurx-2.5.1/Makefile.in	2009-11-03 08:57:37.000000000 +0100
@@ -39,6 +39,7 @@
 mandir = @mandir@

 CC = @CC_QUALIFIED@
+AR = @AR@
 CFLAGS = @CFLAGS@ -I ${srcdir}
 LDFLAGS = @LDFLAGS@

@@ -67,8 +68,8 @@
 libgnurx-$(DLLVERSION).dll libgnurx.dll.a: $(OBJECTS)
 	$(CC) $(CFLAGS) -shared -o libgnurx-$(DLLVERSION).dll $(LDFLAGS) $(OBJECTS)

-libregex.a: libgnurx.dll.a
-	cp -p libgnurx.dll.a $@
+libregex.a: $(OBJECTS)
+	$(AR) cru $@ $(OBJECTS)

 gnurx.lib: libgnurx-$(DLLVERSION).dll
