diff -ru ltrace-0.5.3.orig/sysdeps/linux-gnu/Makefile ltrace-0.5.3/sysdeps/linux-gnu/Makefile
--- ltrace-0.5.3.orig/sysdeps/linux-gnu/Makefile	2009-07-25 17:13:02.000000000 +0200
+++ ltrace-0.5.3/sysdeps/linux-gnu/Makefile	2010-01-01 22:20:45.000000000 +0100
@@ -2,6 +2,7 @@
 				-e s/i.86/i386/ \
 				-e s/sun4u/sparc64/ \
 				-e s/sparc64/sparc/ \
+				-e s/alpha.*/alpha/ \
 				-e s/arm.*/arm/ \
 				-e s/sa110/arm/ \
 				-e s/ppc64/ppc/ \
