Index: ltrace-0.5.3/configure
===================================================================
--- ltrace-0.5.3/configure
+++ ltrace-0.5.3/configure
@@ -78,7 +78,10 @@ fi
 rm -f conftest.c a.out
 
 CFLAGS='-g -O2'
-LIBS='-lelf -lsupc++ -liberty '
+LIBS='-lelf'
+[ -n "$HAVE_LIBSUPC__" ] && LIBS="$LIBS -lsupc++"
+[ -n "$HAVE_LIBIBERTY" ] && LIBS="$LIBS -liberty"
+
 INSTALL='install -c'
 iquote='-iquote '
 iquoteend=''
