diff -ru ltrace-0.5.3.org/Makefile.in ltrace-0.5.3/Makefile.in
--- ltrace-0.5.3.org/Makefile.in	2011-08-21 18:55:15.000000000 +0200
+++ ltrace-0.5.3/Makefile.in	2011-08-21 18:40:53.000000000 +0200
@@ -15,6 +15,7 @@
 mandir		= @mandir@
 docdir		= $(prefix)/share/doc/ltrace
 
+AR		=	@AR@
 CC		=	@CC@
 CFLAGS		=	-Wall @CFLAGS@
 CPPFLAGS	=	-iquote $(TOPDIR) -iquote $(TOPDIR)/sysdeps/$(OS) -DSYSCONFDIR=\"$(sysconfdir)\" @CPPFLAGS@

diff -ru ltrace-0.5.3.org/configure ltrace-0.5.3/configure
--- ltrace-0.5.3.org/configure	2011-08-21 18:55:15.000000000 +0200
+++ ltrace-0.5.3/configure	2011-08-21 18:54:46.000000000 +0200
@@ -113,7 +117,7 @@
 #
 # Makefile.in -> Makefile
 #
-x_subst_vars='PACKAGE_VERSION HOST HOST_OS INSTALL CC CPPFLAGS CFLAGS LDFLAGS LIBS iquote iquoteend prefix sysconfdir mandir docdir'
+x_subst_vars='PACKAGE_VERSION HOST HOST_OS INSTALL AR CC CPPFLAGS CFLAGS LDFLAGS LIBS iquote iquoteend prefix sysconfdir mandir docdir'
 
 for i in $x_subst_vars
 do
