diff -urpN newlib-2.1.0.orig/libgloss/arm/configure newlib-2.1.0/libgloss/arm/configure
--- newlib-2.1.0.orig/libgloss/arm/configure	2017-04-11 00:41:24.592873407 -0700
+++ newlib-2.1.0/libgloss/arm/configure	2017-04-11 00:43:12.305788823 -0700
@@ -2541,7 +2541,7 @@ test "${CCASFLAGS+set}" = set || CCASFLA
 
 
 case "${target}" in
-  *-*-elf | *-*-eabi)
+  *-*-elf | *-*-eabi*)
 	objtype=elf-
 	;;
   *-*-coff)
diff -urpN newlib-2.1.0.orig/libgloss/arm/configure.in newlib-2.1.0/libgloss/arm/configure.in
--- newlib-2.1.0.orig/libgloss/arm/configure.in	2017-04-11 00:41:24.592873407 -0700
+++ newlib-2.1.0/libgloss/arm/configure.in	2017-04-11 00:43:12.305788823 -0700
@@ -49,7 +49,7 @@ AC_PROG_RANLIB
 LIB_AM_PROG_AS
 
 case "${target}" in
-  *-*-elf | *-*-eabi)
+  *-*-elf | *-*-eabi*)
 	objtype=elf-
 	;;
   *-*-coff)
