Index: gcc-4.9.0/gcc/config/microblaze/microblaze.h
===================================================================
--- gcc-4.9.0.orig/gcc/config/microblaze/microblaze.h
+++ gcc-4.9.0/gcc/config/microblaze/microblaze.h
@@ -218,6 +218,12 @@ extern enum pipeline_type microblaze_pip
 #undef PTRDIFF_TYPE
 #define PTRDIFF_TYPE "int"
 
+#undef SIZE_TYPE
+#define SIZE_TYPE "unsigned int"
+
+#undef PTRDIFF_TYPE
+#define PTRDIFF_TYPE "int"
+
 #define CONSTANT_ALIGNMENT(EXP, ALIGN)					\
   ((TREE_CODE (EXP) == STRING_CST  || TREE_CODE (EXP) == CONSTRUCTOR)	\
    && (ALIGN) < BITS_PER_WORD						\
