--- /dev/null	Sat Dec 14 13:56:51 2002
+++ gcc-3.3.1/gcc/testsuite/gcc.dg/pr10589-1.c	Mon Sep 15 08:57:58 2003
@@ -0,0 +1,14 @@
+/* PR target/10589
+ * Reporter: mathieu@thenesis.com
+ * Summary: For Hitachi SH target, GCC crashes when both -fomit-frame-pointer and -mdalign options are specified
+ * Keywords: ice-on-valid-code
+ * Testcase by Kazu Hirata, tweaked by dank@kegel.com
+ * Did not fail in Hard Hat 2.0 gcc-2.97, nor in dodes gcc-3.0.2, so this is a regression, so I'm not marking it xfail
+ */
+/* { dg-do compile { target sh*-*-* } } */
+/* { dg-options "-fomit-frame-pointer -mdalign" } */
+
+int foo(int a, int b)
+{
+	return a / b;
+}
