--- /dev/null	Sat Dec 14 13:56:51 2002
+++ gcc-3.3.1/gcc/testsuite/gcc.dg/pr11587-1.c	Mon Sep 15 08:46:00 2003
@@ -0,0 +1,36 @@
+/* PR optimization/11587
+ * Reporter: Michael Eager <eager@mvista.com>
+ * Summary: [3.3/3.4 Regression] SH ICE in reload_cse_simplify_operand, postreload.c
+ * Keywords: ice-on-valid-code
+ *
+ * Note: a fix exists; see bugzilla
+ *
+ * Testcase tweaked by dank@kegel.com
+ * Not marked as xfail since it's a regression.
+ */
+
+/* { dg-do compile } */
+/* { dg-options "-O1" } */
+int foo1(void);
+int foo2();
+int foo3();
+
+static int goo()
+{
+	int i;
+
+	if (i <= 0)
+		return i;
+
+	if (foo1() > 1) {
+		i = foo2();
+		if (i < 0)
+			return i;
+		if (i) {
+			if (foo3())
+				return 1;
+			return 0;
+		}
+	}
+
+}
