--- /dev/null	Sat Dec 14 13:56:51 2002
+++ gcc-3.3.1/gcc/testsuite/gcc.dg/pr11162-1.c	Sun Sep 14 16:24:41 2003
@@ -0,0 +1,22 @@
+/*
+ * PR optimization/11162
+ * Reporter: Toshiyasu Morita <tm@kloo.net>
+ * Summary: [3.4 Regression] [-fnew-ra] ICE compiling channel.i on sh
+ * Keywords: ice-on-valid-code
+ * Description:
+ * [ICE] when channel.i from stress-1.17 is compiled with options:
+ * -O2 -m4 -fnew-ra
+ * ...
+ * The CFG code gets confused by the addition of a USE insn after a sibcall.
+ * Sibcalls are supposed to end the block (and the function!), so the fact
+ * that we have an instruction following one (even if its fake) is confusing.
+ *
+ * testcase tweaked by dank@kegel.com
+ */
+/* { dg-do compile } */
+/* { dg-options "-O2 -fnew-ra" } */
+
+int foo(char *p1)
+{
+	return bar(p1);
+}
