From 0f4f6c1cc9888c9637d5e9304661888597c5dde1 Mon Sep 17 00:00:00 2001
From: Chris Packham <judge.packham@gmail.com>
Date: Mon, 8 Aug 2022 20:31:10 +1200
Subject: [PATCH 0/5] sh conf

---
 configure    | 2 +-
 configure.ac | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/configure b/configure
index bbfb5979..bcd95684 100755
--- a/configure
+++ b/configure
@@ -4056,7 +4056,7 @@ case "${target}" in
   nvptx*-*-*)
     noconfigdirs="$noconfigdirs target-libssp target-libstdc++-v3 target-libobjc"
     ;;
-  sh-*-*)
+  sh*-*-*)
     case "${target}" in
       sh*-*-elf)
          ;;
diff --git a/configure.ac b/configure.ac
index 659388f7..eb1c20af 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1278,7 +1278,7 @@ case "${target}" in
   nvptx*-*-*)
     noconfigdirs="$noconfigdirs target-libssp target-libstdc++-v3 target-libobjc"
     ;;
-  sh-*-*)
+  sh*-*-*)
     case "${target}" in
       sh*-*-elf)
          ;;
-- 
2.50.1

