Index: sysdeps/powerpc/powerpc32/configure.in
===================================================================
--- libc/sysdeps/powerpc/powerpc32/configure.in	(revision 7000)
+++ libc/sysdeps/powerpc/powerpc32/configure.in	(working copy)
@@ -1,6 +1,8 @@
 GLIBC_PROVIDES dnl See aclocal.m4 in the top level source directory.
 # Local configure fragment for sysdeps/powerpc/powerpc32.
 
+AC_PREREQ(2.50)
+
 # See whether gas has R_PPC_REL16 relocs.
 AC_CACHE_CHECK(for R_PPC_REL16 gas support, libc_cv_ppc_rel16, [dnl
 cat > conftest.s <<\EOF
Index: sysdeps/powerpc/powerpc32/elf/configure.in
===================================================================
--- libc/sysdeps/powerpc/powerpc32/elf/configure.in	(revision 7000)
+++ libc/sysdeps/powerpc/powerpc32/elf/configure.in	(working copy)
@@ -1,6 +1,8 @@
 GLIBC_PROVIDES dnl See aclocal.m4 in the top level source directory.
 # Local configure fragment for sysdeps/powerpc32/elf.
 
+AC_PREREQ(2.50)
+
 if test "$usetls" != no; then
 # Check for support of thread-local storage handling in assembler and
 # linker.
Index: sysdeps/powerpc/configure.in
===================================================================
--- libc/sysdeps/powerpc/configure.in	(revision 7000)
+++ libc/sysdeps/powerpc/configure.in	(working copy)
@@ -1,6 +1,8 @@
 GLIBC_PROVIDES dnl See aclocal.m4 in the top level source directory.
 # Local configure fragment for sysdeps/powerpc.
 
+AC_PREREQ(2.50)
+
 # Accept binutils which knows about ".machine".
 AC_CACHE_CHECK(for .machine support, libc_cv_ppc_machine, [dnl
 cat > conftest.s <<\EOF
Index: sysdeps/unix/sysv/linux/powerpc/configure.in
===================================================================
--- libc/sysdeps/unix/sysv/linux/powerpc/configure.in	(revision 7000)
+++ libc/sysdeps/unix/sysv/linux/powerpc/configure.in	(working copy)
@@ -2,6 +2,8 @@
 GLIBC_PROVIDES dnl See aclocal.m4 in the top level source directory.
 # Local configure fragment for sysdeps/unix/sysv/linux/powerpc/.
 
+AC_PREREQ(2.50)
+
 AC_CACHE_CHECK(whether $CC $CFLAGS -mlong-double-128 uses IBM extended format,
 	       libc_cv_mlong_double_128ibm, [dnl
 save_CFLAGS="$CFLAGS"
Index: sysdeps/ieee754/ldbl-opt/configure.in
===================================================================
--- libc/sysdeps/ieee754/ldbl-opt/configure.in	(revision 7000)
+++ libc/sysdeps/ieee754/ldbl-opt/configure.in	(working copy)
@@ -2,6 +2,8 @@
 GLIBC_PROVIDES dnl See aclocal.m4 in the top level source directory.
 # Local configure fragment for sysdeps/ieee754/ldbl-opt/.
 
+AC_PREREQ(2.50)
+
 AC_CACHE_CHECK(whether $CC $CFLAGS supports -mlong-double-128,
 	       libc_cv_mlong_double_128, [dnl
 save_CFLAGS="$CFLAGS"
Index: libidn/configure.in
===================================================================
--- libc/libidn/configure.in	(revision 7000)
+++ libc/libidn/configure.in	(working copy)
@@ -1,6 +1,8 @@
 dnl glibc configure fragment for libidn add-on
 GLIBC_PROVIDES dnl See aclocal.m4 in the top level source directory.
 
+AC_PREREQ(2.50)
+
 libc_add_on_canonical=
 libc_add_on_subdirs=.
 
