diff -durN ltrace-0.5.2.orig/configure ltrace-0.5.2/configure
--- ltrace-0.5.2.orig/configure	2009-09-14 18:47:45.000000000 +0200
+++ ltrace-0.5.2/configure	2009-09-14 18:49:18.000000000 +0200
@@ -1829,6 +1829,9 @@
   linux-gnu*) host_os=linux-gnu
 esac
 HOST_OS="$host_os"
+case "${HOST_OS}" in
+  linux-*) HOST_OS=linux-gnu;;
+esac
 
 
 ac_ext=c
diff -durN ltrace-0.5.2.orig/configure.ac ltrace-0.5.2/configure.ac
--- ltrace-0.5.2.orig/configure.ac	2009-09-14 18:47:45.000000000 +0200
+++ ltrace-0.5.2/configure.ac	2009-09-14 18:49:18.000000000 +0200
@@ -10,6 +10,9 @@
   linux-gnu*) host_os=linux-gnu
 esac
 HOST_OS="$host_os"
+case "${HOST_OS}" in
+  linux-*) HOST_OS=linux-gnu;;
+esac
 AC_SUBST(HOST_OS)
 HOST_ARCH="$host_cpu"
 AC_SUBST(HOST_ARCH)
