diff -dur uClibc-0.9.28.1.orig/libc/sysdeps/linux/common/Makefile uClibc-0.9.28.1/libc/sysdeps/linux/common/Makefile
--- uClibc-0.9.28.1.orig/libc/sysdeps/linux/common/Makefile	2007-01-26 00:54:19.000000000 +0100
+++ uClibc-0.9.28.1/libc/sysdeps/linux/common/Makefile	2007-02-01 08:44:37.000000000 +0100
@@ -25,6 +25,10 @@
 SRCS := $(filter-out sbrk.c,$(SRCS))
 endif
 
+ifneq ($(strip $(UCLIBC_HAS_LFS)),y)
+SRCS := $(filter-out readahead.c,$(SRCS))
+endif
+
 ifneq ($(strip $(UCLIBC_HAS_SSP)),y)
 SRCS := $(filter-out ssp.c,$(SRCS))
 endif
