--- gcc-4.1.2.orig/libstdc++-v3/include/ext/rope	2005-08-17 04:28:44.000000000 +0200
+++ gcc-4.1.2/libstdc++-v3/include/ext/rope	2007-03-04 23:34:08.000000000 +0100
@@ -57,6 +57,9 @@
 #include <bits/allocator.h>
 #include <ext/hash_fun.h>
 
+/* cope w/ index defined as macro, SuSv3 proposal */
+#undef index
+
 # ifdef __GC
 #   define __GC_CONST const
 # else
--- gcc-4.1.2.orig/libstdc++-v3/include/ext/ropeimpl.h	2005-08-17 04:28:44.000000000 +0200
+++ gcc-4.1.2/libstdc++-v3/include/ext/ropeimpl.h	2007-03-04 23:34:08.000000000 +0100
@@ -53,6 +53,9 @@
 #include <ext/memory> // For uninitialized_copy_n
 #include <ext/numeric> // For power
 
+/* cope w/ index defined as macro, SuSv3 proposal */
+#undef index
+
 namespace __gnu_cxx
 {
   using std::size_t;
