--- gcc-4.1.2.orig/gcc/configure.ac	2006-11-13 23:09:55.000000000 +0100
+++ gcc-4.1.2/gcc/configure.ac	2007-03-04 23:34:53.000000000 +0100
@@ -2435,7 +2435,7 @@
 	tls_first_minor=14
 	tls_as_opt="-m64 -Aesame --fatal-warnings"
 	;;
-  sh-*-* | sh[34]-*-*)
+  sh-*-* | sh[34]*-*-*)
     conftest_s='
 	.section ".tdata","awT",@progbits
 foo:	.long	25
--- gcc-4.1.2.orig/gcc/configure	2006-11-13 23:09:55.000000000 +0100
+++ gcc-4.1.2/gcc/configure	2007-03-04 23:34:53.000000000 +0100
@@ -14762,7 +14762,7 @@
 	tls_first_minor=14
 	tls_as_opt="-m64 -Aesame --fatal-warnings"
 	;;
-  sh-*-* | sh[34]-*-*)
+  sh-*-* | sh[34]*-*-*)
     conftest_s='
 	.section ".tdata","awT",@progbits
 foo:	.long	25
