--- gcc-4.1.2.orig/libstdc++-v3/configure	2007-03-04 23:32:31.000000000 +0100
+++ gcc-4.1.2/libstdc++-v3/configure	2007-03-04 23:33:54.000000000 +0100
@@ -7324,6 +7324,9 @@
 cat >>conftest.$ac_ext <<_ACEOF
 /* end confdefs.h.  */
 #include <complex.h>
+#ifdef __UCLIBC__
+#error ugly hack to make sure configure test fails here for cross until uClibc supports the complex funcs
+#endif
 int
 main ()
 {
