diff -ur gcc-6.2.0.orig/gcc/config/nios2/nios2.h gcc-6.2.0/gcc/config/nios2/nios2.h
--- gcc-6.2.0.orig/gcc/config/nios2/nios2.h	2016-11-29 10:27:50.364479625 -0800
+++ gcc-6.2.0/gcc/config/nios2/nios2.h	2016-11-29 10:29:55.069624746 -0800
@@ -63,11 +63,11 @@
 #if TARGET_ENDIAN_DEFAULT == 0
 # define ASM_SPEC "%{!meb:-EL} %{meb:-EB} %{march=*:-march=%*}"
 # define LINK_SPEC_ENDIAN "%{!meb:-EL} %{meb:-EB}"
-# define MULTILIB_DEFAULTS { "EL" }
+# define MULTILIB_DEFAULTS { "mel" }
 #else
 # define ASM_SPEC "%{!mel:-EB} %{mel:-EL} %{march=*:-march=%*}"
 # define LINK_SPEC_ENDIAN "%{!mel:-EB} %{mel:-EL}"
-# define MULTILIB_DEFAULTS { "EB" }
+# define MULTILIB_DEFAULTS { "meb" }
 #endif
 
 #define LINK_SPEC LINK_SPEC_ENDIAN \
diff -ur gcc-6.2.0.orig/gcc/config/nios2/t-nios2 gcc-6.2.0/gcc/config/nios2/t-nios2
--- gcc-6.2.0.orig/gcc/config/nios2/t-nios2	2016-11-29 10:27:50.364479625 -0800
+++ gcc-6.2.0/gcc/config/nios2/t-nios2	2016-11-29 10:29:03.517151014 -0800
@@ -22,6 +22,5 @@
 # MULTILIB_DIRNAMES = nomul mulx fpu-60-1 fpu-60-2
 # MULTILIB_EXCEPTIONS = 
 
-# MULTILIB_OPTIONS += EL/EB
+# MULTILIB_OPTIONS += mel/meb
 # MULTILIB_DIRNAMES += le be
-# MULTILIB_MATCHES += EL=mel EB=meb
