From 07e515506660b1d0c1934dc0ac0e2ac5e7a4a760 Mon Sep 17 00:00:00 2001
From: "Dmitry V. Levin" <ldv@altlinux.org>
Date: Wed, 22 Aug 2012 12:58:18 -0700
Subject: [PATCH] Fix sed configure check for newer sed --version output.

---
diff --git a/configure b/configure
index 9dddf3c..a573bfe 100755
--- a/configure
+++ b/configure
@@ -4963,7 +4963,7 @@ else
   # Found it, now check the version.
   { $as_echo "$as_me:${as_lineno-$LINENO}: checking version of $SED" >&5
 $as_echo_n "checking version of $SED... " >&6; }
-  ac_prog_version=`$SED --version 2>&1 | sed -n 's/^.*GNU sed version \([0-9]*\.[0-9.]*\).*$/\1/p'`
+  ac_prog_version=`$SED --version 2>&1 | sed -n 's/^.*GNU sed[^0-9]* \([0-9]*\.[0-9.]*\).*$/\1/p'`
   case $ac_prog_version in
     '') ac_prog_version="v. ?.??, bad"; ac_verc_fail=yes;;
     3.0[2-9]*|3.[1-9]*|[4-9]*)
diff --git a/configure.in b/configure.in
index f5dbff6..1b05c87 100644
--- a/configure.in
+++ b/configure.in
@@ -909,7 +909,7 @@ AC_CHECK_PROG_VER(MAKEINFO, makeinfo, --version,
   [4.[5-9]*|4.[1-9][0-9]*|[5-9].*],
   MAKEINFO=: aux_missing="$aux_missing makeinfo")
 AC_CHECK_PROG_VER(SED, sed, --version,
-  [GNU sed version \([0-9]*\.[0-9.]*\)],
+  [GNU sed[^0-9]* \([0-9]*\.[0-9.]*\)],
   [3.0[2-9]*|3.[1-9]*|[4-9]*],
   SED=: aux_missing="$aux_missing sed")
 
-- 
1.9.4

