diff -Naurd cloog-ppl-0.15.7-a/configure cloog-ppl-0.15.7-b/configure
--- cloog-ppl-0.15.7-a/configure	2009-08-12 03:33:31.000000000 +0200
+++ cloog-ppl-0.15.7-b/configure	2012-01-27 13:26:13.970988501 +0100
@@ -12902,7 +12902,7 @@
 
 rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
 
-	LIBS="$LIBS -lppl_c -lppl -lgmpxx"
+	LIBS="-lppl_c -lppl -lgmpxx $LIBS"
 
 cat >>confdefs.h <<\_ACEOF
 #define CLOOG_PPL_BACKEND 1
diff -Naurd cloog-ppl-0.15.7-a/configure.in cloog-ppl-0.15.7-b/configure.in
--- cloog-ppl-0.15.7-a/configure.in	2009-08-12 03:33:31.000000000 +0200
+++ cloog-ppl-0.15.7-b/configure.in	2012-01-27 13:26:13.970988501 +0100
@@ -345,7 +345,7 @@
 	#endif
 	], [AC_MSG_RESULT([yes])], [AC_MSG_ERROR(Can't find correct version of PPL.) ])
 
-	LIBS="$LIBS -lppl_c -lppl -lgmpxx"
+	LIBS="-lppl_c -lppl -lgmpxx $LIBS"
 	AC_DEFINE([CLOOG_PPL_BACKEND], 1, [Use the PPL backend])
 
 else
