diff -urpN gcc-5.3.0.orig/gcc/config/nios2/nios2.h gcc-5.3.0/gcc/config/nios2/nios2.h
--- gcc-5.3.0.orig/gcc/config/nios2/nios2.h	2016-04-18 10:32:41.046254061 -0700
+++ gcc-5.3.0/gcc/config/nios2/nios2.h	2016-04-18 10:37:25.998841674 -0700
@@ -55,11 +55,11 @@
 #if TARGET_ENDIAN_DEFAULT == 0
 # define ASM_SPEC "%{!meb:-EL} %{meb:-EB}"
 # define LINK_SPEC_ENDIAN "%{!meb:-EL} %{meb:-EB}"
-# define MULTILIB_DEFAULTS { "EL" }
+# define MULTILIB_DEFAULTS { "mel" }
 #else
 # define ASM_SPEC "%{!mel:-EB} %{mel:-EL}"
 # define LINK_SPEC_ENDIAN "%{!mel:-EB} %{mel:-EL}"
-# define MULTILIB_DEFAULTS { "EB" }
+# define MULTILIB_DEFAULTS { "meb" }
 #endif
 
 #define LINK_SPEC LINK_SPEC_ENDIAN \
diff -urpN gcc-5.3.0.orig/gcc/config/nios2/t-nios2 gcc-5.3.0/gcc/config/nios2/t-nios2
--- gcc-5.3.0.orig/gcc/config/nios2/t-nios2	2016-04-18 10:32:41.046254061 -0700
+++ gcc-5.3.0/gcc/config/nios2/t-nios2	2016-04-18 10:36:41.091312157 -0700
@@ -22,6 +22,5 @@
 # MULTILIB_DIRNAMES = nomul mulx fpu-60-1 fpu-60-2
 # MULTILIB_EXCEPTIONS = 
 
-# MULTILIB_OPTIONS += EL/EB
+# MULTILIB_OPTIONS += mel/meb
 # MULTILIB_DIRNAMES += le be
-# MULTILIB_MATCHES += EL=mel EB=meb
