From bdb18461e4d78d701c7d4a43667b55af6eb2ad13 Mon Sep 17 00:00:00 2001
From: Chris Packham <judge.packham@gmail.com>
Date: Mon, 8 Aug 2022 20:31:10 +1200
Subject: [PATCH 0/5] sh conf

---
 configure    | 2 +-
 configure.ac | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/configure b/configure
index 2551fe02..4d50547e 100755
--- a/configure
+++ b/configure
@@ -4196,7 +4196,7 @@ case "${target}" in
   nvptx*-*-*)
     noconfigdirs="$noconfigdirs target-libssp"
     ;;
-  sh-*-*)
+  sh*-*-*)
     case "${target}" in
       sh*-*-elf)
          ;;
diff --git a/configure.ac b/configure.ac
index fe7c1cb4..e5f1e828 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1365,7 +1365,7 @@ case "${target}" in
   nvptx*-*-*)
     noconfigdirs="$noconfigdirs target-libssp"
     ;;
-  sh-*-*)
+  sh*-*-*)
     case "${target}" in
       sh*-*-elf)
          ;;
-- 
2.53.0

