Index: binutils-2.19.51.0.5/ld/ld.texinfo
===================================================================

diff -durN binutils-2.20.1.orig/ld/ld.texinfo binutils-2.20.1/ld/ld.texinfo
--- binutils-2.20.1.orig/ld/ld.texinfo	2009-07-06 15:48:51.000000000 +0200
+++ binutils-2.20.1/ld/ld.texinfo	2010-08-17 19:32:16.000000000 +0200
@@ -2026,8 +2026,9 @@
 systems may not understand them. If you specify
 @option{--enable-new-dtags}, the dynamic tags will be created as needed.
 If you specify @option{--disable-new-dtags}, no new dynamic tags will be
-created. By default, the new dynamic tags are not created. Note that
-those options are only available for ELF systems.
+created. On Gentoo, by default, the new dynamic tags are created (this
+differs from upstream behaviour). Note that those options are only
+available for ELF systems.
 
 @kindex --hash-size=@var{number}
 @item --hash-size=@var{number}
