diff -ur glibc-2.13/ports/sysdeps/arm/preconfigure glibc-2.13-patched/ports/sysdeps/arm/preconfigure
--- glibc-2.13/ports/sysdeps/arm/preconfigure	2011-01-25 14:00:16.000000000 -0700
+++ glibc-2.13-patched/ports/sysdeps/arm/preconfigure	2017-03-07 15:36:50.310454000 -0700
@@ -2,7 +2,7 @@
 arm*)
 	base_machine=arm
 	case $config_os in
-	linux-gnueabi)
+	linux-gnueabi*)
 		machine=arm/eabi/$machine
 		if [ "${CFLAGS+set}" != "set" ]; then
 		  CFLAGS="-g -O2"
diff -ur glibc-2.13/ports/sysdeps/arm/shlib-versions glibc-2.13-patched/ports/sysdeps/arm/shlib-versions
--- glibc-2.13/ports/sysdeps/arm/shlib-versions	2011-01-25 14:00:16.000000000 -0700
+++ glibc-2.13-patched/ports/sysdeps/arm/shlib-versions	2017-03-07 15:36:50.312457000 -0700
@@ -1,4 +1,4 @@
-arm.*-.*-linux-gnueabi	DEFAULT			GLIBC_2.4
+arm.*-.*-linux-gnueabi.*	DEFAULT			GLIBC_2.4
 
-arm.*-.*-linux-gnueabi	ld=ld-linux.so.3
+arm.*-.*-linux-gnueabi.*	ld=ld-linux.so.3
 arm.*-.*-linux.*	ld=ld-linux.so.2
