diff --git a/stdlib/setenv.c b/stdlib/setenv.c
index da61ee0..72cf877 100644
--- a/stdlib/setenv.c
+++ b/stdlib/setenv.c
@@ -277,7 +277,7 @@ unsetenv (const char *name)
 
   ep = __environ;
   if (ep != NULL)
-    while (*ep != NULL)
+    while (*ep != NULL){
       if (!strncmp (*ep, name, len) && (*ep)[len] == '=')
 	{
 	  /* Found it.  Remove this pointer by moving later ones back.  */
@@ -290,6 +290,7 @@ unsetenv (const char *name)
 	}
       else
 	++ep;
+    }
 
   UNLOCK;

diff --git a/nis/nis_call.c b/nis/nis_call.c
index 3fa37e4..1b7bb47 100644
--- a/nis/nis_call.c
+++ b/nis/nis_call.c
@@ -680,7 +680,7 @@ nis_server_cache_add (const_nis_name name, int search_parent,
   /* Choose which entry should be evicted from the cache.  */
   loc = &nis_server_cache[0];
   if (*loc != NULL)
-    for (i = 1; i < 16; ++i)
+    for (i = 1; i < 16; ++i){
       if (nis_server_cache[i] == NULL)
 	{
 	  loc = &nis_server_cache[i];
@@ -690,6 +690,7 @@ nis_server_cache_add (const_nis_name name, int search_parent,
 	       || ((*loc)->uses == nis_server_cache[i]->uses
 		   && (*loc)->expires > nis_server_cache[i]->expires))
 	loc = &nis_server_cache[i];
+    }
   old = *loc;
   *loc = new;
 
