From 0d37b5114a1caa2e1cb8e5273b6a674abee37b91 Mon Sep 17 00:00:00 2001
From: Chris Packham <judge.packham@gmail.com>
Date: Mon, 8 Aug 2022 20:31:10 +1200
Subject: [PATCH] sh conf

---
 configure    |    2 +-
 configure.ac |    2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

--- a/configure
+++ b/configure
@@ -3911,7 +3911,7 @@
   nvptx*-*-*)
     noconfigdirs="$noconfigdirs target-libssp target-libstdc++-v3 target-libobjc"
     ;;
-  sh-*-*)
+  sh*-*-*)
     case "${target}" in
       sh*-*-elf)
          ;;
--- a/configure.ac
+++ b/configure.ac
@@ -1189,7 +1189,7 @@
   nvptx*-*-*)
     noconfigdirs="$noconfigdirs target-libssp target-libstdc++-v3 target-libobjc"
     ;;
-  sh-*-*)
+  sh*-*-*)
     case "${target}" in
       sh*-*-elf)
          ;;
