--- glibc-2.2.5/configure.old	2003-05-30 21:51:13.000000000 -0700
+++ glibc-2.2.5/configure	2003-05-30 21:51:23.000000000 -0700
@@ -1738,7 +1738,7 @@
   ac_cv_prog_cc_works=yes
   # If we can't run a trivial program, we are probably using a cross compiler.
   if (./conftest; exit) 2>/dev/null; then
-    ac_cv_prog_cc_cross=no
+    ac_cv_prog_cc_cross=yes
   else
     ac_cv_prog_cc_cross=yes
   fi
