diff -up ./tests/Makefile.am.orig ./tests/Makefile.am
--- ./tests/Makefile.am.orig	2009-07-28 16:28:55.377059021 +0300
+++ ./tests/Makefile.am	2009-07-28 16:31:01.136810734 +0300
@@ -20,7 +20,7 @@ check_PROGRAMS = tversion tinternals tin
 
 EXTRA_DIST = tgeneric.c tgeneric_ui.c mpf_compat.h inp_str.data tmul.dat
 
-tmul_CPPFLAGS = -DMPFR_SRCDIR=$(srcdir)
+tmul_CPPFLAGS = -DMPFR_SRCDIR=\"$(srcdir)\"
 
 LDADD = libfrtests.la $(MPFR_LIBM) $(top_builddir)/libmpfr.la
 INCLUDES = -I$(top_srcdir) -I$(top_builddir)
diff -up ./tests/tmul.c.orig ./tests/tmul.c
--- ./tests/tmul.c.orig	2009-07-28 16:29:03.761058919 +0300
+++ ./tests/tmul.c	2009-07-28 16:30:46.369812215 +0300
@@ -26,10 +26,8 @@ MA 02110-1301, USA. */
 #include "mpfr-test.h"
 
 #ifndef MPFR_SRCDIR
-#define MPFR_SRCDIR .
+#define MPFR_SRCDIR "."
 #endif
-#define QUOTE(X) NAME(X)
-#define NAME(X) #X
 
 #ifdef CHECK_EXTERNAL
 static int
@@ -482,10 +480,10 @@ check_regression (void)
   mpfr_inits2 (6177, x, y, z, (mpfr_ptr) 0);
   /* we read long strings from a file since ISO C90 does not support strings of
      length > 509 */
-  fp = fopen (QUOTE (MPFR_SRCDIR)"/tmul.dat", "r");
+  fp = fopen (MPFR_SRCDIR "/tmul.dat", "r");
   if (fp == NULL)
     {
-      fprintf (stderr, "Error, cannot open "QUOTE (MPFR_SRCDIR)"/tmul.dat\n");
+      fprintf (stderr, "Error, cannot open " MPFR_SRCDIR "/tmul.dat\n");
       exit (1);
     }
   get_string (s, fp);
