diff -dur duma_2_5_1/Makefile duma_2_5_1.orig/Makefile
--- duma_2_5_1/Makefile	2006-06-18 06:02:13.000000000 +0200
+++ duma_2_5_1.orig/Makefile	2007-07-11 17:49:43.000000000 +0200
@@ -98,10 +98,6 @@
   LIBS=-lpthread
 endif
 
-
-CC=gcc
-CXX=g++
-AR=ar
 INSTALL=install
 
 ############################################################
@@ -169,7 +165,7 @@
 
 createconf: createconf.o
 	- rm -f createconf
-	$(CC) $(CFLAGS) $(DUMA_OPTIONS) createconf.o -o createconf
+	$(HOSTCC) $(HOSTCFLAGS) $(DUMA_OPTIONS) createconf.o -o createconf
 
 tstheap: libduma.a tstheap.o
 	- rm -f tstheap
@@ -211,7 +207,7 @@
 # define rules how to build objects for createconf
 #
 createconf.o:
-	$(CC) $(CFLAGS) $(DUMA_OPTIONS) -c createconf.c -o $@
+	$(HOSTCC) $(HOSTCFLAGS) $(DUMA_OPTIONS) -c createconf.c -o $@
 
 
 #
