diff -dur duma_2_5_1.orig/Makefile duma_2_5_1/Makefile
--- duma_2_5_1.orig/Makefile	2007-07-11 17:49:43.000000000 +0200
+++ duma_2_5_1/Makefile	2007-07-11 17:50:46.000000000 +0200
@@ -115,9 +115,14 @@
 	dumatest.c tstheap.c dumatestpp.cpp testoperators.cpp \
 	createconf.c
 
-OBJECTS = dumapp.o duma.o sem_inc.o print.o
+OBJECTS = duma.o sem_inc.o print.o
+SO_OBJECTS = duma_so.o sem_inc_so.o print_so.o
+
+ifeq ($(DUMA_CPP),1)
+OBJECTS += dumapp.o
+SO_OBJECTS += dumapp_so.o
+endif
 
-SO_OBJECTS = dumapp_so.o duma_so.o sem_inc_so.o print_so.o
 
 all:	libduma.a $(DUMASO) tstheap dumatest dumatestpp testoperators tstheap_so dumatestpp_so
 	@ echo "Testing DUMA."
