From 3b16131609b2edbf29194532c089079a8d2664b1 Mon Sep 17 00:00:00 2001
From: Jakub Labenski <kuba@parasoft.com>
Date: Fri, 5 Jun 2020 09:11:40 +0200
Subject: [PATCH] Fix GCC 10+ detection

---
 configure | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure b/configure
index 1ee4c42003..4e1d1fa8a0 100755
--- a/configure
+++ b/configure
@@ -4709,7 +4709,7 @@ $as_echo_n "checking version of $CC... " >&6; }
   ac_prog_version=`$CC -v 2>&1 | sed -n 's/^.*version \([egcygnustpi-]*[0-9.]*\).*$/\1/p'`
   case $ac_prog_version in
     '') ac_prog_version="v. ?.??, bad"; ac_verc_fail=yes;;
-    4.[4-9].* | 4.[1-9][0-9].* | [5-9].* )
+    4.[4-9].* | 4.[1-9][0-9].* | [5-9].* | [1-9][0-9].* )
        ac_prog_version="$ac_prog_version, ok"; ac_verc_fail=no;;
     *) ac_prog_version="$ac_prog_version, bad"; ac_verc_fail=yes;;
 
-- 
2.25.1

