diff -Naurd mpfr-2.4.1-a/PATCHES mpfr-2.4.1-b/PATCHES
--- mpfr-2.4.1-a/PATCHES	2009-03-04 13:15:05.000000000 +0000
+++ mpfr-2.4.1-b/PATCHES	2009-03-04 13:17:04.000000000 +0000
@@ -0,0 +1 @@
+assert
diff -Naurd mpfr-2.4.1-a/VERSION mpfr-2.4.1-b/VERSION
--- mpfr-2.4.1-a/VERSION	2009-02-27 16:55:37.000000000 +0000
+++ mpfr-2.4.1-b/VERSION	2009-03-04 13:16:46.000000000 +0000
@@ -1 +1 @@
-2.4.1-p1
+2.4.1-p2
diff -Naurd mpfr-2.4.1-a/mpfr-gmp.c mpfr-2.4.1-b/mpfr-gmp.c
--- mpfr-2.4.1-a/mpfr-gmp.c	2009-02-20 09:43:17.000000000 +0000
+++ mpfr-2.4.1-b/mpfr-gmp.c	2009-03-04 13:16:39.000000000 +0000
@@ -301,7 +301,7 @@
       if (linenum != -1)
         fprintf (stderr, "%d: ", linenum);
     }
-  fprintf (stderr, " assertion failed: %s\n", expr);
+  fprintf (stderr, "MPFR assertion failed: %s\n", expr);
   abort();
 }
 
diff -Naurd mpfr-2.4.1-a/mpfr.h mpfr-2.4.1-b/mpfr.h
--- mpfr-2.4.1-a/mpfr.h	2009-02-27 16:55:38.000000000 +0000
+++ mpfr-2.4.1-b/mpfr.h	2009-03-04 13:16:46.000000000 +0000
@@ -27,7 +27,7 @@
 #define MPFR_VERSION_MAJOR 2
 #define MPFR_VERSION_MINOR 4
 #define MPFR_VERSION_PATCHLEVEL 1
-#define MPFR_VERSION_STRING "2.4.1-p1"
+#define MPFR_VERSION_STRING "2.4.1-p2"
 
 /* Macros dealing with MPFR VERSION */
 #define MPFR_VERSION_NUM(a,b,c) (((a) << 16L) | ((b) << 8) | (c))
diff -Naurd mpfr-2.4.1-a/version.c mpfr-2.4.1-b/version.c
--- mpfr-2.4.1-a/version.c	2009-02-27 16:55:38.000000000 +0000
+++ mpfr-2.4.1-b/version.c	2009-03-04 13:16:46.000000000 +0000
@@ -25,5 +25,5 @@
 const char *
 mpfr_get_version (void)
 {
-  return "2.4.1-p1";
+  return "2.4.1-p2";
 }
