diff -Naurd mpfr-2.4.1-a/tests/tfprintf.c mpfr-2.4.1-b/tests/tfprintf.c
--- mpfr-2.4.1-a/tests/tfprintf.c	2009-02-20 09:43:15.000000000 +0000
+++ mpfr-2.4.1-b/tests/tfprintf.c	2009-03-10 01:12:45.000000000 +0000
@@ -195,7 +195,7 @@
   check_vfprintf (fout, "a. %Pu, b. %c, c. %Zi%Zn", prec, ch, mpz, &mpz);
   check_length_with_cmp (6, mpz, 17, mpz_cmp_ui (mpz, 17), Zi);
   check_vfprintf (fout, "%% a. %#.0RNg, b. %Qx%Rn, c. %p", mpfr, mpq, &mpfr,
-                  &i);
+                  (void *) &i);
   check_length_with_cmp (7, mpfr, 16, mpfr_cmp_ui (mpfr, 16), Rg);
 
 #ifndef NPRINTF_T
diff -Naurd mpfr-2.4.1-a/tests/tprintf.c mpfr-2.4.1-b/tests/tprintf.c
--- mpfr-2.4.1-a/tests/tprintf.c	2009-02-20 09:43:15.000000000 +0000
+++ mpfr-2.4.1-b/tests/tprintf.c	2009-03-10 01:12:45.000000000 +0000
@@ -191,7 +191,8 @@
   check_length (5, sz, 34, zu);
   check_vprintf ("a. %Pu, b. %c, c. %RUG, d. %Zi%Zn", prec, ch, mpfr, mpz, &mpz);
   check_length_with_cmp (6, mpz, 24, mpz_cmp_ui (mpz, 24), Zi);
-  check_vprintf ("%% a. %#.0RNg, b. %Qx%Rn c. %p", mpfr, mpq, &mpfr, &i);
+  check_vprintf ("%% a. %#.0RNg, b. %Qx%Rn c. %p",
+                 mpfr, mpq, &mpfr, (void *) &i);
   check_length_with_cmp (7, mpfr, 16, mpfr_cmp_ui (mpfr, 16), Rg);
 
 #ifndef NPRINTF_T
