diff -ru a/config.guess b/config.guess
--- a/config.guess	2021-05-14 10:42:08.000000000 +0200
+++ b/config.guess	2021-12-20 01:38:22.000000000 +0100
@@ -1276,6 +1276,9 @@
     *:Rhapsody:*:*)
 	echo ${UNAME_MACHINE}-apple-rhapsody${UNAME_RELEASE}
 	exit ;;
+    arm64:Darwin:*:*)
+	echo aarch64-apple-darwin"$UNAME_RELEASE"
+	exit ;;
     *:Darwin:*:*)
 	UNAME_PROCESSOR=`uname -p` || UNAME_PROCESSOR=unknown
 	eval $set_cc_for_build
