Index: gcc-4.8.1/libgcc/unwind-dw2-fde-dip.c
===================================================================
--- gcc-4.8.1.orig/libgcc/unwind-dw2-fde-dip.c
+++ gcc-4.8.1/libgcc/unwind-dw2-fde-dip.c
@@ -73,6 +73,13 @@
     && defined(TARGET_DL_ITERATE_PHDR) \
     && defined(__sun__) && defined(__svr4__)
 # define USE_PT_GNU_EH_FRAME
+ #endif
+
+/* For musl libc, TARGET_DL_ITERATE_PHDR gets defined by the configure
+   script. */
+#if !defined(inhibit_libc) && defined(HAVE_LD_EH_FRAME_HDR)	\
+  && defined(TARGET_DL_ITERATE_PHDR)
+# define USE_PT_GNU_EH_FRAME
 #endif
 
 #if defined(USE_PT_GNU_EH_FRAME)
