diff -durN binutils-2.22.orig/ld/Makefile.am binutils-2.22/ld/Makefile.am
--- binutils-2.22.orig/ld/Makefile.am	2011-07-22 22:22:37.000000000 +0200
+++ binutils-2.22/ld/Makefile.am	2011-12-14 19:50:25.760779164 +0100
@@ -37,7 +37,7 @@
 # We put the scripts in the directory $(scriptdir)/ldscripts.
 # We can't put the scripts in $(datadir) because the SEARCH_DIR
 # directives need to be different for native and cross linkers.
-scriptdir = $(tooldir)/lib
+scriptdir = $(libdir)
 
 EMUL = @EMUL@
 EMULATION_OFILES = @EMULATION_OFILES@
diff -durN binutils-2.22.orig/ld/Makefile.in binutils-2.22/ld/Makefile.in
--- binutils-2.22.orig/ld/Makefile.in	2011-07-22 22:22:37.000000000 +0200
+++ binutils-2.22/ld/Makefile.in	2011-12-14 19:50:25.784779163 +0100
@@ -366,7 +366,7 @@
 # We put the scripts in the directory $(scriptdir)/ldscripts.
 # We can't put the scripts in $(datadir) because the SEARCH_DIR
 # directives need to be different for native and cross linkers.
-scriptdir = $(tooldir)/lib
+scriptdir = $(libdir)
 BASEDIR = $(srcdir)/..
 BFDDIR = $(BASEDIR)/bfd
 INCDIR = $(BASEDIR)/include
