diff -Naurd cloog-ppl-0.15.6-a/configure cloog-ppl-0.15.6-b/configure
--- cloog-ppl-0.15.6-a/configure	2009-08-10 23:18:07.000000000 +0200
+++ cloog-ppl-0.15.6-b/configure	2012-01-27 13:25:08.858988857 +0100
@@ -12917,7 +12917,7 @@
 
 rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
 
-	LIBS="$LIBS -lppl_c -lppl -lgmpxx"
+	LIBS="-lppl_c -lppl -lgmpxx $LIBS"
 
 cat >>confdefs.h <<\_ACEOF
 #define CLOOG_PPL_BACKEND 1
diff -Naurd cloog-ppl-0.15.6-a/configure.in cloog-ppl-0.15.6-b/configure.in
--- cloog-ppl-0.15.6-a/configure.in	2009-08-10 23:18:07.000000000 +0200
+++ cloog-ppl-0.15.6-b/configure.in	2012-01-27 13:25:08.862988846 +0100
@@ -354,7 +354,7 @@
 	#endif
 	], [AC_MSG_RESULT([yes])], [AC_MSG_ERROR(Can't find correct version of PPL.) ])
 
-	LIBS="$LIBS -lppl_c -lppl -lgmpxx"
+	LIBS="-lppl_c -lppl -lgmpxx $LIBS"
 	AC_DEFINE([CLOOG_PPL_BACKEND], 1, [Use the PPL backend])
 
 else
