diff -durN glibc-cvs-2.9.orig/misc/syslog.c glibc-cvs-2.9/misc/syslog.c
--- glibc-cvs-2.9.orig/misc/syslog.c	2009-06-01 10:16:50.000000000 +0200
+++ glibc-cvs-2.9/misc/syslog.c	2009-06-01 10:17:20.000000000 +0200
@@ -152,7 +152,7 @@
 #define	INTERNALLOG	LOG_ERR|LOG_CONS|LOG_PERROR|LOG_PID
 	/* Check for invalid bits. */
 	if (pri & ~(LOG_PRIMASK|LOG_FACMASK)) {
-		syslog(INTERNALLOG,
+		__syslog(INTERNALLOG,
 		    "syslog: unknown facility/priority: %x", pri);
 		pri &= LOG_PRIMASK|LOG_FACMASK;
 	}
