diff -ur glibc-2.12.1/ports/sysdeps/arm/preconfigure glibc-2.12.1-patched/ports/sysdeps/arm/preconfigure
--- glibc-2.12.1/ports/sysdeps/arm/preconfigure	2017-03-07 15:33:56.410265000 -0700
+++ glibc-2.12.1-patched/ports/sysdeps/arm/preconfigure	2017-03-07 15:34:42.016840000 -0700
@@ -2,7 +2,7 @@
 arm*)
 	base_machine=arm
 	case $config_os in
-	linux-gnueabi)
+	linux-gnueabi*)
 		machine=arm/eabi/$machine
 		;;
 	*)
diff -ur glibc-2.12.1/ports/sysdeps/arm/shlib-versions glibc-2.12.1-patched/ports/sysdeps/arm/shlib-versions
--- glibc-2.12.1/ports/sysdeps/arm/shlib-versions	2017-03-07 15:33:56.439267000 -0700
+++ glibc-2.12.1-patched/ports/sysdeps/arm/shlib-versions	2017-03-07 15:34:42.018837000 -0700
@@ -1,4 +1,4 @@
-arm.*-.*-linux-gnueabi	DEFAULT			GLIBC_2.4
+arm.*-.*-linux-gnueabi.*	DEFAULT			GLIBC_2.4
 
-arm.*-.*-linux-gnueabi	ld=ld-linux.so.3
+arm.*-.*-linux-gnueabi.*	ld=ld-linux.so.3
 arm.*-.*-linux.*	ld=ld-linux.so.2
