--- strace-orig/file.c	2008/05/20 04:56:18	1.92
+++ strace/file.c	2008/08/01 01:13:10	1.93
@@ -28,21 +28,22 @@
  * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
  * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  *
- *	$Id: file.c,v 1.92 2008/05/20 04:56:18 roland Exp $
+ *	$Id: file.c,v 1.93 2008/08/01 01:13:10 roland Exp $
  */
 
 #include "defs.h"
 
 #include <dirent.h>
+
 #ifdef LINUX
-#define dirent kernel_dirent
-#define dirent64 kernel_dirent64
-#include <linux/types.h>
-#include <linux/dirent.h>
-#undef dirent
-#undef dirent64
+struct kernel_dirent {
+	unsigned long   d_ino;
+	unsigned long   d_off;
+	unsigned short  d_reclen;
+	char            d_name[1];
+};
 #else
-#define kernel_dirent dirent
+# define kernel_dirent dirent
 #endif
 
 #ifdef LINUX
