diff -durN binutils-2.18.50.0.8.orig/configure binutils-2.18.50.0.8/configure
--- binutils-2.18.50.0.8.orig/configure	2008-07-10 17:32:56.000000000 +0200
+++ binutils-2.18.50.0.8/configure	2008-07-13 23:08:13.000000000 +0200
@@ -2313,7 +2313,7 @@
     noconfigdirs="$noconfigdirs target-libffi target-qthreads"
     libgloss_dir=arm
     ;;
-  arm*-*-linux-gnueabi)
+  arm*-*-linux-gnueabi | arm*-*-linux-uclibcgnueabi)
     noconfigdirs="$noconfigdirs target-qthreads"
     case ${with_newlib} in
       no) noconfigdirs="$noconfigdirs target-newlib target-libgloss"
diff -durN binutils-2.18.50.0.8.orig/configure.ac binutils-2.18.50.0.8/configure.ac
--- binutils-2.18.50.0.8.orig/configure.ac	2008-07-10 17:32:56.000000000 +0200
+++ binutils-2.18.50.0.8/configure.ac	2008-07-13 23:08:13.000000000 +0200
@@ -562,7 +562,7 @@
     noconfigdirs="$noconfigdirs target-libffi target-qthreads"
     libgloss_dir=arm
     ;;
-  arm*-*-linux-gnueabi)
+  arm*-*-linux-gnueabi | arm*-*-linux-uclibcgnueabi)
     noconfigdirs="$noconfigdirs target-qthreads"
     case ${with_newlib} in
       no) noconfigdirs="$noconfigdirs target-newlib target-libgloss"
