From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ahmed ARIF <arif.ing@outlook.com>
Date: Mon, 12 Jan 2026 00:00:00 +0000
Subject: [PATCH] mingw-w64: add fmal support for AArch64

Provide a basic implementation of fmal for AArch64 Windows.

--- a/mingw-w64-crt/math/fmal.c
+++ b/mingw-w64-crt/math/fmal.c
@@ -12,6 +12,13 @@
 /* On ARM `long double` is 64 bits. And ARM has hardware FMA. */
 long double fmal(long double x, long double y, long double z){
   return fma(x, y, z);
+}
+
+#elif defined(__aarch64__)
+
+long double fmal(long double x, long double y, long double z)
+{
+  return x * y + z;
 }

 #elif defined(_AMD64_) || defined(__x86_64__) || defined(_X86_) || defined(__i386__)
