diff -durN gcc-4.4.5.orig/gcc/config/alpha/alpha.h gcc-4.4.5/gcc/config/alpha/alpha.h
--- gcc-4.4.5.orig/gcc/config/alpha/alpha.h	2009-02-20 16:20:38.000000000 +0100
+++ gcc-4.4.5/gcc/config/alpha/alpha.h	2010-10-09 22:57:31.000000000 +0200
@@ -95,6 +95,8 @@
   while (0)
 #endif
 
+#define CPP_SPEC "%{!no-ieee:-mieee}"
+
 #define WORD_SWITCH_TAKES_ARG(STR)		\
  (!strcmp (STR, "rpath") || DEFAULT_WORD_SWITCH_TAKES_ARG(STR))
 
diff -durN gcc-4.4.5.orig/gcc/config/alpha/alpha.opt gcc-4.4.5/gcc/config/alpha/alpha.opt
--- gcc-4.4.5.orig/gcc/config/alpha/alpha.opt	2007-08-02 12:49:31.000000000 +0200
+++ gcc-4.4.5/gcc/config/alpha/alpha.opt	2010-10-09 22:57:31.000000000 +0200
@@ -39,7 +39,7 @@
 Request IEEE-conformant math library routines (OSF/1)
 
 mieee
-Target Report RejectNegative Mask(IEEE)
+Target Report Mask(IEEE)
 Emit IEEE-conformant code, without inexact exceptions
 
 mieee-with-inexact
