diff -durN gcc-4.4.5.orig/gcc/configure gcc-4.4.5/gcc/configure
--- gcc-4.4.5.orig/gcc/configure	2010-06-20 17:43:53.000000000 +0200
+++ gcc-4.4.5/gcc/configure	2010-10-09 22:58:36.000000000 +0200
@@ -13997,7 +13997,7 @@
 	    | powerpc*-*-*,powerpc64*-*-*)
 		CROSS="$CROSS -DNATIVE_CROSS" ;;
 	esac
-elif test "x$TARGET_SYSTEM_ROOT" != x; then
+elif test "x$TARGET_SYSTEM_ROOT" != x -o $build != $host; then
         SYSTEM_HEADER_DIR=$build_system_header_dir
 fi
 
diff -durN gcc-4.4.5.orig/gcc/configure.ac gcc-4.4.5/gcc/configure.ac
--- gcc-4.4.5.orig/gcc/configure.ac	2010-06-20 17:43:53.000000000 +0200
+++ gcc-4.4.5/gcc/configure.ac	2010-10-09 22:58:36.000000000 +0200
@@ -1720,7 +1720,7 @@
 	    | powerpc*-*-*,powerpc64*-*-*)
 		CROSS="$CROSS -DNATIVE_CROSS" ;;
 	esac
-elif test "x$TARGET_SYSTEM_ROOT" != x; then
+elif test "x$TARGET_SYSTEM_ROOT" != x -o $build != $host; then
         SYSTEM_HEADER_DIR=$build_system_header_dir 
 fi
 
diff -durN gcc-4.4.5.orig/gcc/unwind-dw2.c gcc-4.4.5/gcc/unwind-dw2.c
--- gcc-4.4.5.orig/gcc/unwind-dw2.c	2010-01-20 09:39:18.000000000 +0100
+++ gcc-4.4.5/gcc/unwind-dw2.c	2010-10-09 22:58:36.000000000 +0200
@@ -329,9 +329,11 @@
 }
 #endif
 
+#ifndef inhibit_libc
 #ifdef MD_UNWIND_SUPPORT
 #include MD_UNWIND_SUPPORT
 #endif
+#endif
 
 /* Extract any interesting information from the CIE for the translation
    unit F belongs to.  Return a pointer to the byte after the augmentation,
