diff -durN gcc-4.4.5.orig/gcc/config/netbsd-elf.h gcc-4.4.5/gcc/config/netbsd-elf.h
--- gcc-4.4.5.orig/gcc/config/netbsd-elf.h	2007-09-03 18:14:04.000000000 +0200
+++ gcc-4.4.5/gcc/config/netbsd-elf.h	2010-10-09 22:59:57.000000000 +0200
@@ -82,6 +82,7 @@
 #define NETBSD_LINK_SPEC_ELF \
   "%{assert*} %{R*} %{rpath*} \
    %{shared:-shared} \
+   %{symbolic:-Bsymbolic} \
    %{!shared: \
      -dc -dp \
      %{!nostdlib: \
