diff -durN gcc-4.4.5.orig/include/libiberty.h gcc-4.4.5/include/libiberty.h
--- gcc-4.4.5.orig/include/libiberty.h	2008-06-24 03:42:31.000000000 +0200
+++ gcc-4.4.5/include/libiberty.h	2010-10-09 23:01:30.000000000 +0200
@@ -595,8 +595,11 @@
 /* Like sprintf but provides a pointer to malloc'd storage, which must
    be freed by the caller.  */
 
+/* asprintf may be declared as a macro by glibc with __USE_FORTIFY_LEVEL.  */
+#ifndef asprintf
 extern int asprintf (char **, const char *, ...) ATTRIBUTE_PRINTF_2;
 #endif
+#endif
 
 #if !HAVE_DECL_VASPRINTF
 /* Like vsprintf but provides a pointer to malloc'd storage, which
