diff -durN gcc-4.4.5.orig/gcc/configure gcc-4.4.5/gcc/configure
--- gcc-4.4.5.orig/gcc/configure	2010-10-09 22:58:36.000000000 +0200
+++ gcc-4.4.5/gcc/configure	2010-10-09 23:04:50.000000000 +0200
@@ -22285,7 +22285,7 @@
 	tls_first_minor=14
 	tls_as_opt="-m64 -Aesame --fatal-warnings"
 	;;
-  sh-*-* | sh[34]-*-*)
+  sh-*-* | sh[34]*-*-*)
     conftest_s='
 	.section ".tdata","awT",@progbits
 foo:	.long	25
diff -durN gcc-4.4.5.orig/gcc/configure.ac gcc-4.4.5/gcc/configure.ac
--- gcc-4.4.5.orig/gcc/configure.ac	2010-10-09 22:58:36.000000000 +0200
+++ gcc-4.4.5/gcc/configure.ac	2010-10-09 23:04:50.000000000 +0200
@@ -2728,7 +2728,7 @@
 	tls_first_minor=14
 	tls_as_opt="-m64 -Aesame --fatal-warnings"
 	;;
-  sh-*-* | sh[34]-*-*)
+  sh-*-* | sh[34]*-*-*)
     conftest_s='
 	.section ".tdata","awT",@progbits
 foo:	.long	25
