From aad130ddadd19ae40e5405a7b9ddd5ae1ec138f0 Mon Sep 17 00:00:00 2001
From: Chris Packham <judge.packham@gmail.com>
Date: Mon, 8 Aug 2022 20:36:31 +1200
Subject: [PATCH 1/8] check_ldrunpath_length

---
 ld/ldelf.c |    4 ++++
 1 file changed, 4 insertions(+)

--- a/ld/ldelf.c
+++ b/ld/ldelf.c
@@ -1130,6 +1130,8 @@
 		  && command_line.rpath == NULL)
 		{
 		  path = (const char *) getenv ("LD_RUN_PATH");
+		  if (path && *path == '\0')
+		    path = NULL;
 		  if (path
 		      && ldelf_search_needed (path, &n, force,
 					      is_linux, elfsize))
@@ -1805,6 +1807,8 @@
   rpath = command_line.rpath;
   if (rpath == NULL)
     rpath = (const char *) getenv ("LD_RUN_PATH");
+  if (rpath && *rpath == '\0')
+    rpath = NULL;
 
   for (abfd = link_info.input_bfds; abfd; abfd = abfd->link.next)
     if (bfd_get_flavour (abfd) == bfd_target_elf_flavour)
