--- a/gcc/config/moxie/moxiebox.h	2018-11-07 08:52:40.481629768 +0100
+++ b/gcc/config/moxie/moxiebox.h	2018-01-03 11:03:58.000000000 +0100
@@ -31,7 +31,7 @@
 #undef LIB_SPEC
 #define LIB_SPEC \
 "%{!T*:-Tmoxiebox.ld} \
- %{!nostdlib: --start-group -lsandboxrt -lc -lgcc --end-group }"
+ %{!nostdlib: --start-group -lc -lgcc --end-group }"
 
 #undef LINK_SPEC
 #define LINK_SPEC "%{h*} %{v:-V} -EL -Bstatic"
