diff -durN binutils-2.21.1.orig/configure binutils-2.21.1/configure
--- binutils-2.21.1.orig/configure	2011-06-27 09:39:00.000000000 +0200
+++ binutils-2.21.1/configure	2011-09-12 13:39:06.292251293 +0200
@@ -3180,7 +3180,7 @@
     noconfigdirs="$noconfigdirs target-libffi target-qthreads"
     libgloss_dir=arm
     ;;
-  arm*-*-linux-gnueabi)
+  arm*-*-linux-*gnueabi)
     noconfigdirs="$noconfigdirs target-qthreads"
     case ${with_newlib} in
       no) noconfigdirs="$noconfigdirs target-newlib target-libgloss"
diff -durN binutils-2.21.1.orig/configure.ac binutils-2.21.1/configure.ac
--- binutils-2.21.1.orig/configure.ac	2011-06-27 11:08:53.000000000 +0200
+++ binutils-2.21.1/configure.ac	2011-09-12 13:39:06.308917895 +0200
@@ -652,7 +652,7 @@
     noconfigdirs="$noconfigdirs target-libffi target-qthreads"
     libgloss_dir=arm
     ;;
-  arm*-*-linux-gnueabi)
+  arm*-*-linux-*gnueabi)
     noconfigdirs="$noconfigdirs target-qthreads"
     case ${with_newlib} in
       no) noconfigdirs="$noconfigdirs target-newlib target-libgloss"
