--- glibc-2.6/sysdeps/unix/sysv/linux/kernel-features.h	2006-07-29 07:43:27.000000000 +0300
+++ glibc-2.6/sysdeps/unix/sysv/linux/kernel-features.h.new	2008-07-26 08:45:06.000000000 +0300
@@ -445,7 +445,7 @@
 /* The *at syscalls were introduced just after 2.6.16-rc1.  Due to the way the
    kernel versions are advertised we can only rely on 2.6.17 to have
    the code.  */
-#if __LINUX_KERNEL_VERSION >= 0x020611
+#if __LINUX_KERNEL_VERSION >= 0x020611 && !defined __alpha__
 # define __ASSUME_ATFCTS	1
 #endif
 
