diff -dur binutils-2.17.old/bfd/configure binutils-2.17/bfd/configure
--- binutils-2.17.old/bfd/configure	2006-06-23 20:17:03.000000000 +0200
+++ binutils-2.17/bfd/configure	2007-05-01 18:26:03.000000000 +0200
@@ -3576,7 +3576,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux-gnu*|linux-uclibc*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
diff -dur binutils-2.17.old/binutils/configure binutils-2.17/binutils/configure
--- binutils-2.17.old/binutils/configure	2006-04-06 23:49:29.000000000 +0200
+++ binutils-2.17/binutils/configure	2007-05-01 18:26:03.000000000 +0200
@@ -3411,7 +3411,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux-gnu*|linux-uclibc*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
diff -dur binutils-2.17.old/configure binutils-2.17/configure
--- binutils-2.17.old/configure	2006-04-06 23:49:25.000000000 +0200
+++ binutils-2.17/configure	2007-05-01 18:26:03.000000000 +0200
@@ -1276,7 +1276,7 @@
   am33_2.0-*-linux*)
     noconfigdirs="$noconfigdirs ${libgcj} target-newlib target-libgloss"
     ;;
-  sh-*-linux*)
+  sh*-*-linux*)
     noconfigdirs="$noconfigdirs ${libgcj} target-newlib target-libgloss"
     ;;    
   sh*-*-pe|mips*-*-pe|*arm-wince-pe)
@@ -1584,7 +1584,7 @@
   romp-*-*)
     noconfigdirs="$noconfigdirs bfd binutils ld gas opcodes target-libgloss ${libgcj}"
     ;;
-  sh-*-* | sh64-*-*)
+  sh*-*-* | sh64-*-*)
     case "${host}" in
       i[3456789]86-*-vsta) ;; # don't add gprof back in
       i[3456789]86-*-go32*) ;; # don't add gprof back in
diff -dur binutils-2.17.old/configure.in binutils-2.17/configure.in
--- binutils-2.17.old/configure.in	2006-06-23 20:19:53.000000000 +0200
+++ binutils-2.17/configure.in	2007-05-01 18:26:03.000000000 +0200
@@ -468,7 +468,7 @@
   am33_2.0-*-linux*)
     noconfigdirs="$noconfigdirs ${libgcj} target-newlib target-libgloss"
     ;;
-  sh-*-linux*)
+  sh*-*-linux*)
     noconfigdirs="$noconfigdirs ${libgcj} target-newlib target-libgloss"
     ;;    
   sh*-*-pe|mips*-*-pe|*arm-wince-pe)
@@ -776,7 +776,7 @@
   romp-*-*)
     noconfigdirs="$noconfigdirs bfd binutils ld gas opcodes target-libgloss ${libgcj}"
     ;;
-  sh-*-* | sh64-*-*)
+  sh*-*-* | sh64-*-*)
     case "${host}" in
       i[[3456789]]86-*-vsta) ;; # don't add gprof back in
       i[[3456789]]86-*-go32*) ;; # don't add gprof back in
diff -dur binutils-2.17.old/gas/configure binutils-2.17/gas/configure
--- binutils-2.17.old/gas/configure	2006-04-06 23:49:31.000000000 +0200
+++ binutils-2.17/gas/configure	2007-05-01 18:26:03.000000000 +0200
@@ -3411,7 +3411,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux-gnu*|linux-uclibc*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
diff -dur binutils-2.17.old/gprof/configure binutils-2.17/gprof/configure
--- binutils-2.17.old/gprof/configure	2006-04-06 23:49:33.000000000 +0200
+++ binutils-2.17/gprof/configure	2007-05-01 18:26:03.000000000 +0200
@@ -3414,6 +3414,11 @@
   lt_cv_deplibs_check_method=pass_all
   ;;
 
+linux-uclibc*)
+  lt_cv_deplibs_check_method=pass_all
+  lt_cv_file_magic_test_file=`echo /lib/libuClibc-*.so`
+  ;;
+
 netbsd* | knetbsd*-gnu)
   if echo __ELF__ | $CC -E - | grep __ELF__ > /dev/null; then
     lt_cv_deplibs_check_method='match_pattern /lib[^/\.]+\.so\.[0-9]+\.[0-9]+$'
diff -dur binutils-2.17.old/ld/configure binutils-2.17/ld/configure
--- binutils-2.17.old/ld/configure	2006-04-11 12:36:26.000000000 +0200
+++ binutils-2.17/ld/configure	2007-05-01 18:26:03.000000000 +0200
@@ -3413,7 +3413,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux-gnu*|linux-uclibc*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
diff -dur binutils-2.17.old/libtool.m4 binutils-2.17/libtool.m4
--- binutils-2.17.old/libtool.m4	2005-12-27 17:37:57.000000000 +0100
+++ binutils-2.17/libtool.m4	2007-05-01 18:26:03.000000000 +0200
@@ -739,7 +739,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux-gnu*|linux-uclibc*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
diff -dur binutils-2.17.old/ltconfig binutils-2.17/ltconfig
--- binutils-2.17.old/ltconfig	2005-07-16 04:36:38.000000000 +0200
+++ binutils-2.17/ltconfig	2007-05-01 18:26:03.000000000 +0200
@@ -602,6 +602,7 @@
 
 # Transform linux* to *-*-linux-gnu*, to support old configure scripts.
 case $host_os in
+linux-uclibc*) ;;
 linux-gnu*) ;;
 linux*) host=`echo $host | sed 's/^\(.*-.*-linux\)\(.*\)$/\1-gnu\2/'`
 esac
@@ -1247,7 +1248,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux-gnu*|linux-uclibc*)
   version_type=linux
   need_lib_prefix=no
   need_version=no
diff -dur binutils-2.17.old/opcodes/configure binutils-2.17/opcodes/configure
--- binutils-2.17.old/opcodes/configure	2006-02-17 15:36:28.000000000 +0100
+++ binutils-2.17/opcodes/configure	2007-05-01 18:26:03.000000000 +0200
@@ -3579,7 +3579,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux-gnu*|linux-uclibc*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
