diff -dur binutils-2.17.old/configure binutils-2.17/configure
--- binutils-2.17.old/configure	2007-05-01 18:26:03.000000000 +0200
+++ binutils-2.17/configure	2007-05-01 18:26:06.000000000 +0200
@@ -1305,7 +1305,7 @@
   arm-*-elf* | strongarm-*-elf* | xscale-*-elf* | arm*-*-eabi* )
     noconfigdirs="$noconfigdirs target-libffi target-qthreads"
     ;;
-  arm*-*-linux-gnueabi)
+  arm*-*-linux-gnueabi | arm*-*-linux-uclibcgnueabi)
     noconfigdirs="$noconfigdirs target-libffi target-qthreads"
     noconfigdirs="$noconfigdirs target-libjava target-libobjc"
     ;;
diff -dur binutils-2.17.old/configure.in binutils-2.17/configure.in
--- binutils-2.17.old/configure.in	2007-05-01 18:26:03.000000000 +0200
+++ binutils-2.17/configure.in	2007-05-01 18:26:06.000000000 +0200
@@ -497,7 +497,7 @@
   arm-*-elf* | strongarm-*-elf* | xscale-*-elf* | arm*-*-eabi* )
     noconfigdirs="$noconfigdirs target-libffi target-qthreads"
     ;;
-  arm*-*-linux-gnueabi)
+  arm*-*-linux-gnueabi | arm*-*-linux-uclibcgnueabi)
     noconfigdirs="$noconfigdirs target-libffi target-qthreads"
     noconfigdirs="$noconfigdirs target-libjava target-libobjc"
     ;;
