--- a/m4/ac_check_gmp.m4~	2011-02-27 04:07:47.000000000 -0500
+++ b/m4/ac_check_gmp.m4	2013-07-10 19:57:35.894457259 -0400
@@ -32,6 +32,9 @@
     with_libgmpxx_prefix="$with_libgmp_prefix"
   fi
 fi
+CPPFLAGS="$CPPFLAGS -I$with_libgmp_prefix/include"
+CXXFLAGS="$CXXFLAGS -I$with_libgmpxx_prefix/include"
+LDFLAGS="$CPPFLAGS -L$with_libgmpxx_prefix/lib -L$with_libgmp_prefix/lib"
 
 AC_ARG_WITH(gmp-build,
   AS_HELP_STRING([--with-gmp-build=DIR],
--- a/configure~	2011-02-27 04:07:47.000000000 -0500
+++ b/configure		2013-07-10 19:57:35.894457259 -0400
@@ -10217,6 +10217,9 @@
     with_libgmpxx_prefix="$with_libgmp_prefix"
   fi
 fi
+CPPFLAGS="$CPPFLAGS -I$with_libgmp_prefix/include"
+CXXFLAGS="$CXXFLAGS -I$with_libgmpxx_prefix/include"
+LDFLAGS="$CPPFLAGS -L$with_libgmpxx_prefix/lib -L$with_libgmp_prefix/lib"
 
 
 # Check whether --with-gmp-build was given.
